---
author: "Joe Colgan"
title: "**Transcriptomic analysis of fat bodies of control and inseminated _Bombus terrestris_ workers and queens**"
output:
  pdf_document: default
  html_document: default
fig_width: 4
fig_height: 4
fontsize: 20pt
---

## Introduction
The purpose of the present analysis is to examine similarities and differences in terms of gene expression associated with the fat bodies of _Bombus terrestris_ workers and queens. For the present study, control (unmated), and artificially inseminated bumblebees were collected and fat bodies removed at three different time points (two, four and eight days post-treatment). The study also
included treatment groups to examine changes in gene expression due to artificial insemination alone as well as the dilutant used during the insemination step.  

## Samples:  
The 'fat bodies' dataset consists of 72 samples:  
**Queen samples:** n = 24 
- Control (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)
- Insemination (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)  

**Worker samples:** n = 48 
- Control (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)
- Insemination (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)
- Dilutant (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)
- Nothing (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)  

## Methods  
Cleaned sequences were quality assessed using FastQC ensuring high base quality and 
low adaptor contamination. Cleaned sequences were aligned against the latest 
_Bombus terrestris_ reference genome assembly using [STAR](https://github.com/alexdobin/STAR) with '--quantMode GeneCounts' parameter resulting in the generation of gene-level counts. Such counts were analysed with 
 [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html). Overlap in terms of genes expressed were analysed using upSet. Counts were transformed using variance stabilising transformation and principal component analyses performed using these values. Prior to differential expression analysis, distance matrices were calculated based on Euclidean distance to examine similarities and differences in expression profiles within and across groups. Differential expression analyses were performed using likelihood ratio tests with multiple correction performed using the Benjamini-Hochberg method (adjusted p value < 0.05) to determine global differences between castes, treatment groups and sampling stage (days post treatment). Interaction effects between caste and treatment group were also assessed using a second model. Rank-based Gene Ontology enrichment analysis for differentially expressed genes was performed using a Kolgomorov-Smirnov test implemented in topGO using the 'classic' algorithm and a node size of 20. The 'classic' algorithm was used as opposed to the 'weight01' algorithm to allow for more direct comparisons between sets of terms associated with differentially expressed genes.    
 
## Results  
The first aspect of the analysis involved examining outliers.  
First, the number of expressed genes was compared across individuals using an upset plot (Fig. 1).  
This revealed **8603** genes expressed in all samples (approximately 75% of all genes (n = 11436) expressed in the fat bodies).   
Second, density plots were generated for the percentage of expressed genes per sample for:
- Caste (Fig. 2A)
- Treatment (Fig. 2B)
- Stage (Fig. 2C)  
There was no significant difference in terms of the mean percentage of expressed genes across castes (t-test; p = 0.07) and between treatments (ANOVA; p = 0.18). However, there was a significance difference in percentage of expressed gene between stages (ANOVA; p < 1.2e-11).

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Step One: Load libraries - if not present, install from scratch.
## Similarly, create output directory (if does not exist already)

# Load libraries; install from scratch if needed
libraries <- c("readr",
               "DESeq2",
               "ggplot2",
               "ggpubr",
               "purrr",
               "UpSetR",
               "hash",
               "reshape2",
               "eulerr",
               "knitr",
               "RUVSeq",
               "PoiClaClu",
               "genefilter",
               "pheatmap",
               "RColorBrewer",
               "WGCNA")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE)
    }
}
dir.create("results")
opts_chunk$set(fig.width = 12, fig.height = 8)
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Set paths to folders containing output files from STAR
paths <- list()
paths$sample_batch_info <- "data/sample_information.txt"
paths$htseq_output   <- "input/fat_body/"
# Set relative paths:
paths$htseq_files_relative <- grep(x = list.files(paths$htseq_output,
                                                     recursive = TRUE),
                                      pattern = ".tab",
                                      value   = TRUE)
paths$htseq_files <- file.path(paths$htseq_output,
                               paths$htseq_files_relative)
# Automatically extract file names from kallisto output for colnames
names(paths$htseq_files) <- gsub(paths$htseq_files_relative,
                                    pattern = "/.*",
                                    replacement = "")
for (filenumber in 1:length(paths$htseq_files)) {
  current_name <- names(paths$htseq_files)[filenumber]
  current_file <- paths$htseq_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Step Three: Put sample names and treatments into a samples
## dataframe for DESeqDataSetFromTximport

#Differential expression analysis and data exploration was performed using
# [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html),
#which allows for identfication of genes significantly differentially expressed,
#as well as examination of substructure and clustering through the
#implementation of principal component analysis on normalised gene-level counts.
directory <- "./input/fat_body/"

## Extract sample names and put into df for tximport
samples     <- data.frame(treatment = names(paths$htseq_files))
## Create a list of samples:
sample_files <- grep(".tab",
                    list.files(directory),
                    value = TRUE)

## Read in sample information for room and batch:
samples_info <- read.table(file = paths$sample_batch_info,
                                  header = FALSE,
                                  col.names = c("sample_name",
                                                "caste",
                                                "tissue",
                                                "stage",
                                                "treatment_group"),
                                  row.names = 1)

## Subset samples related to tissue of interest:
samples_info <- subset(x = samples_info,
                       tissue == "Fat_body")

samples_info$sample_name <- row.names(samples_info)
samples_info$file_name   <- row.names(samples_info)

samples_info       <- samples_info[, c(5, 6, 1:4)]
samples_info$stage <- factor(samples_info$stage)

## Rename treatment and stage information:
samples_info$treatment_group <- gsub(pattern = "Insemination_group",
                                     replacement = "AI_Group_1",
                                     samples_info$treatment_group)
samples_info$treatment_group <- gsub(pattern = "Insemination_with_dilutent",
                                     replacement = "AI_Group_2",
                                     samples_info$treatment_group)
samples_info$treatment_group <- gsub(pattern = "Insemination_with_nothing",
                                     replacement = "AI_Group_3",
                                     samples_info$treatment_group)
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
# Step Three: Explore gene-level accounts across samples to identify potential
# outliers.

## Build DESeq2 dataset:
dds_htseq <- DESeqDataSetFromHTSeqCount(sampleTable = samples_info,
                                       directory = directory,
                                       design = ~ caste +
                                                  stage +
                                                  treatment_group)

## Extract and save counts:
dds_htseq_counts <- counts(dds_htseq)
colnames(dds_htseq_counts) <- gsub(pattern = ".ReadsPerGene.out.tab",
                                   replacement = "",
                                   colnames(dds_htseq_counts))
## Save to file:
write.table(x = dds_htseq_counts,
            file = "results/all_fat_body_counts.txt",
            row.names = TRUE,
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")

## Remove low counts:
keep <- rowSums(counts(dds_htseq)) >= 10
dds_htseq <- dds_htseq[keep, ]

## Extract gene-level counts:
counts <- as.data.frame(counts(dds_htseq))
dim(counts)

## Create a list of genes per sample with at least five mapped reads:
new_list <- list()
for (name in 1:ncol(counts)){
  new_name <- paste("B", name, sep = "")
  new_list[[new_name]] <- row.names(subset(x = counts,
                                        counts[name] > 5))
}

## Plot upSet plot and save as a PDF:
pdf(file = "results/upset_plot.pdf",
    height = 12,
    width = 8) # or other device
UpSetR::upset(fromList(new_list),
                  nsets = 40,
                  main.bar.color = "blue",
                  text.scale = 1.1,
                  nintersects = 30,
                  order.by = "freq",
                  mainbar.y.label = "Number of shared genes",
                  sets.x.label = "Genes expressed")
dev.off()
```

```{r figure1, ref.label = "figure1", fig.width = 8, fig.height = 8, echo = FALSE, message=FALSE}
UpSetR::upset(fromList(new_list),
                  nsets = 40,
                  main.bar.color = "blue",
                  text.scale = 1.1,
                  nintersects = 30,
                  order.by = "freq",
                  mainbar.y.label = "Number of shared genes",
                  sets.x.label = "Genes expressed")
```

_**Fig. 1.** Upset plot displaying overlap in gene expression across all samples. For each sample, the number of expressed genes, as well as the overlap with other samples, are provided._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Across all samples, the total number of genes detected:
total_genes <- length(unique(sort(unlist(new_list))))

## Calculate the percentage of expressed genes per sample:
percentages <- list()
for (name in names(new_list)){
  percentages[[name]] <- length(unique(sort(unlist(new_list[[name]])))) /
                                                   total_genes
}

## Examine percentage of expressed genes across:
## - Caste
## - Treatment
## - Stage
percentages_df <- as.data.frame(t(as.data.frame(percentages)))
summary(percentages_df)

samples_info$short_name <- gsub(pattern = ".ReadsPerGene.out.tab",
                                replacement = "",
                                samples_info$sample_name)

row.names(percentages_df) <- gsub(pattern = ".ReadsPerGene.out.tab",
                                  replacement = "",
                                  samples_info$sample_name)

percentages_df$caste <- samples_info[match(row.names(percentages_df),
                                samples_info$short_name), ]$caste

percentages_df$treatment <- samples_info[match(row.names(percentages_df),
                                samples_info$short_name), ]$treatment_group

percentages_df$stage <- samples_info[match(row.names(percentages_df),
                                           samples_info$short_name), ]$stage

colnames(percentages_df)[1] <- "percentage"

## Generate density plot based on caste:
density_caste_plot <- ggplot(data = percentages_df,
                             aes(x = percentage,
                                 colour = caste)) +
  xlab(label = "Percentage of expressed genes") +
  geom_density() +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 10))

## Perform t-test to compare proportion of expressed genes
## between castes:
t.test(subset(percentages_df,
              caste == "Worker")$percentage,
       subset(percentages_df,
              caste == "Queen")$percentage)

## Generate density plot based on treatment:
density_treatment_plot <- ggplot(data = percentages_df,
                                 aes(x = percentage,
                                     colour = treatment)) +
  xlab(label = "Percentage of expressed genes") +
  geom_density() +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 10))

## Perform an ANOVA to compare proportion of expressed genes
## between treatments:
treatment_aov <- aov(percentage ~ treatment,
                     data = percentages_df)
summary(treatment_aov)

## Generate density plot based on stage of ovarian development:
density_stage_plot <- ggplot(data = percentages_df,
                             aes(x = percentage,
                                 colour = stage)) +
  xlab(label = "Percentage of expressed genes") +
  geom_density() +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 10))

## Perform an ANOVA to compare proportion of expressed genes
## between stages:
stage_aov <- aov(percentage ~ stage,
                 data = percentages_df)
summary(stage_aov)
```


```{r figure2, ref.label = "figure2", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a multi-panel plot:
ggarrange(density_caste_plot,
          density_treatment_plot,
          density_stage_plot,
          nrow = 2,
          ncol = 2,
          legend = "none",
          align = "hv",
          labels = c("A",
                     "B",
                     "C"))
```

_**Fig. 2.** Density plots displaying percentage of genes expressed across samples, including: A) caste (red = "queen", blue = "worker); B) treatment (red = "control", blue = "treatment"); and C) stage of ovarian development (blue = "two days", green = "four days", red = "eight days")._  

\newpage  

## Sample clustering.  
To examine general similarities and differences within and between group, a distance matrix based
on variance stabilising transformed (VST) counts was performed.

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Create a DESeq2 object:
deseq_object  <- DESeq(dds_htseq,
                       test = "LRT",
                       reduced = ~caste + stage)

## Transform gene-level counts using variance-stablising transformation:
vsd <- vst(deseq_object,
           blind = FALSE)

## Remove underscore '_':
vsd$stage <- gsub(pattern = "_",
                  replacement = " ",
                  vsd$stage)

## Rename rownames:
vsd$treatment_group <- gsub(pattern = "AI_Group_1",
                                        replacement = "AI Group 1",
                                        vsd$treatment_group)
vsd$treatment_group <- gsub(pattern = "AI_Group_2",
                                        replacement = "AI Group 2",
                                        vsd$treatment_group)
vsd$treatment_group <- gsub(pattern = "AI_Group_3",
                                        replacement = "AI Group 3",
                                        vsd$treatment_group)
## Save object:
saveRDS(object = vsd,
        file = "results/vsd_fat_body_treatment.rds")

## Calculate distance between samples:
sample_dists <- dist(t(assay(vsd)))

## Convert to a matrix:
sample_dist_matrix <- as.matrix(sample_dists)

## Update rownames for plotting:
rownames(sample_dist_matrix) <- paste(vsd$caste,
                                    vsd$treatment_group,
                                    vsd$stage,
                                    sep = " - ")
colnames(sample_dist_matrix) <- NULL

## Set colours for plotting:
colors <- colorRampPalette(rev(brewer.pal(9,
                                          "Blues")))(255)
```

```{r figure3, ref.label = "figure3", fig.width = 8, fig.height = 6, echo = FALSE, message=FALSE}
## Plot distance matrix:
pheatmap(sample_dist_matrix,
         clustering_distance_rows = sample_dists,
         clustering_distance_cols = sample_dists,
         col = colors)
```

_**Fig. 3.** Heatmap displaying distance matrix of genes expressed within and across samples._   

\newpage  

## Principal component analysis.  

To examine further similarities in terms of expression profiles across samples, a principal component analysis using VST gene values was performed. The first principal component, which explains 32% of the variance in the dataset identifies the possibility of outliers with no clear difference explained by the first two components that separate caste. There was evidence that one of the treatments may be driving the pattern (Figs. 4-6). It was decided to rerun the PCA with the two samples separated by PC1 removed to provide more resolution on differences between samples    
```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Here set the two principal components that you want to compare:
## Create function for plotting:
plot_pca_san <- function(object,
                        first_pc = first_pca,
                        second_pc = second_pca,
                        intgroup = "condition",
                        ntop = 1000,
                        returnData = FALSE) {
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
                                                     length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev ^ 2 / sum(pca$sdev ^ 2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup_df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup_df, 1,
                 paste,
                 collapse = " : "))
  } else {
    colData(object)[[intgroup]]
  }
  ## Select the PCAs and percentVar that you like instead of 1 and 2
  d <- data.frame(PC1 = pca$x[, first_pc],
                  PC2 = pca$x[, second_pc],
                  group = group,
                  intgroup_df,
                  name = colData(vsd)[, 1])
  if (returnData) {
    attr(d, "percentVar") <- percentVar[first_pc:second_pc]
    return(d)
  }
    ggplot(data = d,
           aes_string(x = "PC1",
                      y = "PC2",
                      color = "group")) +
            geom_point(data = d,
                       aes(fill = group),
                       colour = "black",
                       pch = 21,
                       size = 3) +
            xlab(paste0("PC", first_pc, ": ",
                        round(percentVar[first_pc] * 100),
                        "%")) +
            ylab(paste0("PC", second_pc, ": ",
                        round(percentVar[second_pc] * 100),
                        "%")) +
            #coord_fixed() +
            #geom_text_repel(size = 3) +
            theme_bw() +
            theme(axis.title = element_text(size = 12,
                                            face = "bold"),
                  axis.text = element_text(size = 10,
                                           face = "plain"),
                  legend.title = element_text(size = 12,
                                              face = "bold"),
                  legend.text = element_text(size = 12,
                                              face = "plain"),
                  legend.position = "top")
}
```

Scatterplots displaying combinations of the first four principal components (PCs) for:  
- Caste (Fig. 4)  
- Treatmnent (Fig. 5)  
- Stage (Fig. 6)  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plots:
## Plot PCs 1 and 2 and colour based on caste:
caste_plot_pc1_2 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("caste"))
caste_plot_pc1_2 <- caste_plot_pc1_2 +
  scale_fill_discrete(name = "Caste")

## Plot PCs 1 and 3 and colour based on caste:
caste_plot_pc1_3 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("caste"))
caste_plot_pc1_3 <- caste_plot_pc1_3 +
  scale_fill_discrete(name = "Caste")

## Plot PCs 1 and 4 and colour based on caste:
caste_plot_pc1_4 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("caste"))
caste_plot_pc1_4 <- caste_plot_pc1_4 +
  scale_fill_discrete(name = "Caste")
```

```{r figure4, ref.label = "figure4", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a multi-panel plot:
ggarrange(caste_plot_pc1_2,
          caste_plot_pc1_3,
          caste_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```

_**Fig. 4.** Principal component analysis with first two principal components shown. Individuals are coloured by caste (queen = "red"; worker = "blue)._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plots:
## Plot PCs 1 and 2 and colour based on treatment:
group_plot_pc1_2 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("treatment_group"))
group_plot_pc1_2 <- group_plot_pc1_2 +
  scale_fill_discrete(name = "Treatment")

## Plot PCs 1 and 3 and colour based on treatment:
group_plot_pc1_3 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("treatment_group"))
group_plot_pc1_3 <- group_plot_pc1_3 +
  scale_fill_discrete(name = "Treatment")

## Plot PCs 1 and 4 and colour based on treatment:
group_plot_pc1_4 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("treatment_group"))
group_plot_pc1_4 <- group_plot_pc1_4 +
  scale_fill_discrete(name = "Treatment")
```

```{r figure5, ref.label = "figure5", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a multi-panel plot:
ggarrange(group_plot_pc1_2,
          group_plot_pc1_3,
          group_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```

_**Fig. 5.** Principal component analysis with first two principal components shown. Individuals are coloured by treatment (control = "red"; inseminated group = "blue)._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plots:
## Plot PCs 1 and 2 and colour based on stage:
stage_plot_pc1_2 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("stage"))
stage_plot_pc1_2 <- stage_plot_pc1_2 +
  scale_fill_discrete(name = "Stage")

## Plot PCs 1 and 3 and colour based on stage:
stage_plot_pc1_3 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("stage"))
stage_plot_pc1_3 <- stage_plot_pc1_3 +
  scale_fill_discrete(name = "Stage")

## Plot PCs 1 and 4 and colour based on stage:
stage_plot_pc1_4 <- plot_pca_san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("stage"))
stage_plot_pc1_4 <- stage_plot_pc1_4 +
  scale_fill_discrete(name = "Stage")
```

```{r figure6, ref.label = "figure6", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a multi-panel plot:
ggarrange(stage_plot_pc1_2,
          stage_plot_pc1_3,
          stage_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv")
```
_**Fig. 6.** Principal component analysis with first two principal components shown. Individuals are coloured by treatment (two days = "blue"; four days = "green"; eight days = "red")_    

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Calculate principal components:
pca <- prcomp(t(assay(vsd)))

## Convert to a dataframe:
pca_df <- as.data.frame(pca$x)

## check the proportion of variance explained by each PC:
summary(pca)
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Identify outliers:
## Based on PCs 1 and 2, anything left of -40 on PC1 is likely an outlier:
samples_to_keep <- subset(x = pca_df,
                          PC1 > -40)
samples_to_keep <- row.names(samples_to_keep)

## Subset samples to keep:
dds_htseq <- dds_htseq[, samples_to_keep]

## Create a DESeq2 object:
deseq_object  <- DESeq(dds_htseq,
                       test = "LRT",
                       reduced = ~caste + stage)

## Transform gene-level counts using variance-stablising transformation:
vsd <- vst(deseq_object,
             blind = FALSE)

## Remove underscore:
vsd$stage <- gsub(pattern = "_",
                  replacement = " ",
                  vsd$stage)

## Rename rownames:
vsd$treatment_group <- gsub(pattern = "AI_Group_1",
                            replacement = "AI Group 1",
                            vsd$treatment_group)

vsd$treatment_group <- gsub(pattern = "AI_Group_2",
                            replacement = "AI Group 2",
                            vsd$treatment_group)

vsd$treatment_group <- gsub(pattern = "AI_Group_3",
                            replacement = "AI Group 3",
                            vsd$treatment_group)

## Calculate distance between samples:
sample_dists <- dist(t(assay(vsd)))

## Convert to a matrix:
sample_dist_matrix <- as.matrix(sample_dists)

## Update row names:
rownames(sample_dist_matrix) <- paste(vsd$caste,
                                      vsd$treatment_group,
                                      vsd$stage,
                                      sep = " - ")
colnames(sample_dist_matrix) <- NULL

## Set colours"
colors <- colorRampPalette(rev(brewer.pal(9,
                                          "Blues")))(255)

## Export heatmap of clustered samples:
pdf(file = "results/fat_body_filtered_pheatmap.pdf",
    height = 10,
    width = 10)
fb_heatmap_filtered <- pheatmap(sample_dist_matrix,
                                   clustering_distance_rows = sample_dists,
                                   clustering_distance_cols = sample_dists,
                                   col = colors)
dev.off()

## Generate plot:
## Plot PCs 1 and 2 and colour based on caste:
caste_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("caste"))
caste_plot_pc1_2 <- caste_plot_pc1_2 +
  scale_fill_discrete(name = "Caste")

## Plot PCs 1 and 3 and colour based on caste:
caste_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("caste"))
caste_plot_pc1_3 <- caste_plot_pc1_3 +
  scale_fill_discrete(name = "Caste")

## Plot PCs 1 and 4 and colour based on caste:
caste_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("caste"))
caste_plot_pc1_4 <- caste_plot_pc1_4 +
  scale_fill_discrete(name = "Caste")
```

```{r figure7, ref.label = "figure7", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a combined plot:
combined_caste_plot <- ggarrange(caste_plot_pc1_2,
                                 caste_plot_pc1_3,
                                 caste_plot_pc1_4,
                                 nrow = 1,
                                 ncol = 3,
                                 align = "hv",
                                 common.legend = TRUE)
```

_**Fig. 7.** Principal component analysis with first two principal components shown. Individuals are coloured by caste (queen = "red"; worker = "blue)._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plot:
## Plot PCs 1 and 2 and colour based on treatment:
group_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("treatment_group"))
group_plot_pc1_2 <- group_plot_pc1_2 +
  scale_fill_brewer(name = "Treatment",
                    palette = "Blues")

## Plot PCs 1 and 3 and colour based on treatment:
group_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("treatment_group"))
group_plot_pc1_3 <- group_plot_pc1_3 +
  scale_fill_brewer(name = "Treatment",
                    palette = "Blues")

## Plot PCs 1 and 4 and colour based on treatment:
group_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("treatment_group"))
group_plot_pc1_4 <- group_plot_pc1_4 +
  scale_fill_brewer(name = "Treatment",
                    palette = "Blues")
```

```{r figure8, ref.label = "figure8", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a combined plot:
combined_group_plot <- ggarrange(group_plot_pc1_2,
                                 group_plot_pc1_3,
                                 group_plot_pc1_4,
                                 nrow = 1,
                                 ncol = 3,
                                 align = "hv",
                                 common.legend = TRUE)
```

_**Fig. 8.** Principal component analysis with first two principal components shown. Individuals are coloured by treatment._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plot:
## Plot PCs 1 and 2 and colour based on stage:
stage_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("stage"))
stage_plot_pc1_2 <- stage_plot_pc1_2 +
  scale_fill_manual(name = "Stage",
                    values = c("black",
                               "grey",
                               "white"))

## Plot PCs 1 and 3 and colour based on stage:
stage_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("stage"))
stage_plot_pc1_3 <- stage_plot_pc1_3 +
  scale_fill_manual(name = "Stage",
                    values = c("black",
                               "grey",
                               "white"))

## Plot PCs 1 and 4 and colour based on stage:
stage_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("stage"))
stage_plot_pc1_4 <- stage_plot_pc1_4 +
  scale_fill_manual(name = "Stage",
                    values = c("black",
                               "grey",
                               "white"))
```

```{r figure9, ref.label = "figure9", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a combined plot:
combined_stage_plot <- ggarrange(stage_plot_pc1_2,
                                 stage_plot_pc1_3,
                                 stage_plot_pc1_4,
                                 nrow = 1,
                                 ncol = 3,
                                 align = "hv",
                                 common.legend = TRUE)
```
_**Fig. 9.** Principal component analysis with first two principal components shown. Individuals are coloured by stage ovarian development_    

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Calculate principal components:
pca <- prcomp(t(assay(vsd)))

## Convert to a dataframe:
pca_df <- as.data.frame(pca$x)

## check the proportion of variance explained by each PC:
summary(pca)
```

Generate a combined plot:

```{r, message = FALSE}
## Generate a combined multi-panel plot based on PCAs:
pca_combined <- ggarrange(combined_caste_plot,
                          combined_group_plot,
                          combined_stage_plot,
                          nrow = 3,
                          ncol = 1,
                          labels = c("a",
                                     "b",
                                     "c"),
                          align = "hv")

## Print to console:
(pca_combined)

## Save to file:
ggsave(filename = "results/fat_body_filtered_pca_plot.pdf",
       height = 8,
       width = 8)
```

Run lintr to check style errors:

```{r, message = FALSE}
lintr::lint(file = "deseq2_analysis_fb_treatment.Rmd")
```