---
author: "Joe Colgan"
title: "**Transcriptomic analysis of brains of control and inseminated _Bombus terrestris_ workers and queens**"
output:
  pdf_document: default
  html_document: default
fig_width: 4
fig_height: 4
fontsize: 20pt
---

## Introduction
The purpose of the present analysis is to examine similarities and differences in terms of gene expression associated with the brain of _Bombus terrestris_ workers and queens. For the present study, control (unmated), and artificially inseminated bumblebees were collected and brains removed at three different time points (two, four and eight days post-ovarian activation). The study also
included treatment groups to examine changes in gene expression due to artificial insemination alone as well as the dilutant used during the insemination step.  

## Samples:  
The 'brain' dataset consists of 72 samples:    
**Queen samples:** n = 24    
- Control (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)  
- Insemination (n = 14; stage 1 n = 4, stage II n = 4, and stage IV n = 4)   

**Worker samples:** n = 48  
- Control (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)  
- Insemination (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)  
- Dilutant (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)  
- Nothing (n = 12; stage 1 n = 4, stage II n = 4, and stage IV n = 4)  

## Methods  
Cleaned sequences were quality assessed using FastQC ensuring high base quality and 
low adaptor contamination. Cleaned sequences were aligned against the latest 
_Bombus terrestris_ reference genome assembly using [STAR](https://github.com/alexdobin/STAR) with '--quantMode GeneCounts' parameter resulting in the generation of gene-level counts. Such counts were analysed with 
 [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html). Overlap in terms of genes expressed were analysed using upSet. Counts were transformed using variance stabilising transformation and principal component analyses performed using these values. Prior to differential expression analysis, distance matrices were calculated based on Euclidean distance to examine similarities and differences in expression profiles within and across groups. Differential expression analyses were performed using likelihood ratio tests with multiple correction performed using the Benjamini-Hochberg method (adjusted p value < 0.05) to determine global differences between castes, treatment groups and sampling stage (days post treatment). Interaction effects between caste and treatment group were also assessed using a second model. Rank-based Gene Ontology enrichment analysis for differentially expressed genes was performed using a Kolgomorov-Smirnov test implemented in topGO using the 'classic' algorithm and a node size of 20. The 'classic' algorithm was used as opposed to the 'weight01' algorithm to allow for more direct comparisons between sets of terms associated with differentially expressed genes.    

\newpage  

## Results  
The first aspect of the analysis involved examining potential outliers (i.e., samples with aberrant expression profiles).  
First, the number of expressed genes was compared across individuals using an upset plot (Fig. 1).  
This revealed **9232** genes expressed in all samples (approximately 81% of all genes (n = 11435) expressed in the brain).   
Second, density plots were generated for the percentage of expressed genes per sample for:
- Caste (Fig. 2A)
- Treatment (Fig. 2B)
- Stage (Fig. 2C)  
There was a significant difference  in terms of the mean percentage of expressed genes across castes (t-test; p = 0.01), but no difference between treatments (ANOVA; p = 0.38) and stage (ANOVA; p = 0.2).  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Step One: Load libraries - if not present, install from scratch.
## Similarly, create output directory (if does not exist already)

# Load libraries; install from scratch if needed
libraries <- c("readr",
               "DESeq2",
               "ggplot2",
               "ggpubr",
               "purrr",
               "UpSetR",
               "hash",
               "reshape2",
               "eulerr",
               "knitr",
               "RUVSeq",
               "PoiClaClu",
               "genefilter",
               "pheatmap",
               "RColorBrewer",
               "WGCNA")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}
dir.create("results")
opts_chunk$set(fig.width = 12, fig.height = 8)
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Set paths to folders containing output files from STAR
paths <- list()
paths$sample_batch_info <- "data/sample_information.txt"
paths$htseq_output   <- "input/brain/"
# Set relative paths:
paths$htseq_files_relative <- grep(x = list.files(paths$htseq_output,
                                                     recursive = TRUE),
                                      pattern = ".tab",
                                      value   = TRUE)
paths$htseq_files <- file.path(paths$htseq_output,
                               paths$htseq_files_relative)
# Automatically extract file names from kallisto output for colnames
names(paths$htseq_files) <- gsub(paths$htseq_files_relative,
                                    pattern = "/.*",
                                    replacement = "")
for (filenumber in 1:length(paths$htseq_files)) {
  current_name <- names(paths$htseq_files)[filenumber]
  current_file <- paths$htseq_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
# Step Three: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

#Differential expression analysis and data exploration was performed using [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html), which allows for identfication of genes significantly differentially expressed, as well as examination of substructure and clustering through the implementation of principal component analysis on normalised gene-level counts.
directory <- "./input/brain/"

## Extract sample names and put into df for tximport
samples     <- data.frame(treatment = names(paths$htseq_files))
## Create a list of samples:
sample_files <- grep(".tab",
                    list.files(directory),
                    value = TRUE)

## Read in sample information for room and batch:
samples_info <- read.table(file = paths$sample_batch_info,
                                  header = FALSE,
                                  col.names = c("sample_name",
                                                "caste",
                                                "tissue",
                                                "stage",
                                                "treatment_group"),
                                  row.names = 1)

## Subset samples related to tissue of interest:
samples_info <- subset(x = samples_info,
                       tissue == "Brain")

samples_info$sample_name <- row.names(samples_info)
samples_info$file_name   <- row.names(samples_info)

samples_info       <- samples_info[, c(5,6, 1:4)]
samples_info$stage <- factor(samples_info$stage)
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
#Step Three: Explore gene-level accounts across samples to identify potential outliers  
#The level of alignment across samples is generally quite good (>75% of all genes detected across samples (Fig. 1) with on average (mean) 87.5% of genes detected across all samples) (Figs.2-4).

## Build DESeq2 dataset:
dds_htseq <- DESeqDataSetFromHTSeqCount(sampleTable = samples_info,
                                       directory = directory,
                                       design = ~ caste +
                                                  stage +
                                                  treatment_group)
## Remove lowly expressed genes:
keep <- rowSums(counts(dds_htseq)) >= 10
dds_htseq <- dds_htseq[keep, ]

## Extract gene-level counts:
counts <- as.data.frame(counts(dds_htseq))
dim(counts)

## Create a list of genes per sample with at least five mapped reads:
new_list <- list()
for (name in 1:ncol(counts)){
  new_name <- paste("B", name, sep = "")
  new_list[[new_name]] <- row.names(subset(x = counts,
                                        counts[name] > 5))
}

## Plot upSet plot and save as a PDF:
pdf(file = "results/brain_upset_plot.pdf",
    height = 12,
    width = 8) # or other device
UpSetR::upset(fromList(new_list), 
                  nsets = 40,
                  main.bar.color = "blue",
                  text.scale = 1.1,
                  nintersects = 30, 
                  order.by = "freq",
                  mainbar.y.label = "Number of shared genes", 
                  sets.x.label = "Genes expressed")
dev.off()
```

```{r figure1, ref.label = "figure1", fig.width = 8, fig.height = 8, echo = FALSE, message=FALSE}
UpSetR::upset(fromList(new_list), 
                  nsets = 40,
                  main.bar.color = "blue",
                  text.scale = 1.1,
                  nintersects = 30, 
                  order.by = "freq",
                  mainbar.y.label = "Number of shared genes", 
                  sets.x.label = "Genes expressed")
```

_**Fig. 1.** Upset plot displaying overlap in gene expression across all samples. For each sample, the number of expressed genes, as well as the overlap with other samples, are provided._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Across all samples, the total number of genes detected:
total_genes <- length(unique(sort(unlist(new_list))))

percentages <- list()
for (name in names(new_list)){
  percentages[[name]] <- length(unique(sort(unlist(new_list[[name]])))) / total_genes
}

## Examine percentage of expressed genes across:
## - Caste
## - Treatment
## - Stage
percentages_df <- as.data.frame(t(as.data.frame(percentages)))
summary(percentages_df)

samples_info$short_name <- gsub(pattern = ".ReadsPerGene.out.tab",
                                replacement = "",
                                samples_info$sample_name)

row.names(percentages_df) <- gsub(pattern = ".ReadsPerGene.out.tab",
                                replacement = "",
                                samples_info$sample_name)

percentages_df$caste <- samples_info[match(row.names(percentages_df),
                                           samples_info$short_name),]$caste

percentages_df$treatment <- samples_info[match(row.names(percentages_df),
                                               samples_info$short_name),]$treatment_group

## Add 'stage' to pre-existing dataframe:
percentages_df$stage <- samples_info[match(row.names(percentages_df),
                                           samples_info$short_name),]$stage

## Update column name:
colnames(percentages_df)[1] <- "percentage"

## Generate density plot:
density_caste_plot <- ggplot(data = percentages_df,
       aes(x = percentage,
           colour = caste)) +
  xlab(label = "Percentage of expressed genes") +
  geom_density() +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 10))

## Check differences in expressed  genes:
t.test(subset(percentages_df,
              caste == "Worker")$percentage,
       subset(percentages_df,
              caste == "Queen")$percentage)

## Generate density plot:
density_treatment_plot <- ggplot(data = percentages_df,
       aes(x = percentage,
           colour = treatment)) +
  xlab(label = "Percentage of expressed genes") +
  geom_density() +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 10))

## Check differences in expressed  genes:
treatment_aov <- aov(percentage ~ treatment,
               data = percentages_df)
summary(treatment_aov)

## Generate density plot:
density_stage_plot <- ggplot(data = percentages_df,
       aes(x = percentage,
           colour = stage)) +
  xlab(label = "Percentage of expressed genes") +
  geom_density() +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 10))

## Check differences in expressed  genes:
stage_aov <- aov(percentage ~ stage,
               data = percentages_df)
summary(stage_aov)
```


```{r figure2, ref.label = "figure2", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Generate a multi-panel plot:
ggarrange(density_caste_plot,
          density_treatment_plot,
          density_stage_plot,
          nrow = 2,
          ncol = 2,
          legend = "none",
          align = "hv",
          labels = c("A",
                     "B",
                     "C"))
```

_**Fig. 2.** Density plots displaying percentage of genes expressed across samples, including: A) caste (red = "queen", blue = "worker); B) treatment (red = "control", blue = "treatment"); and C) stage of ovarian development (red = "two days", green = "four days", blue = "eight days")._  

\newpage  

## Sample clustering.  
To examine general similarities and differences within and between group, a distance matrix based
on variance stabilising transformed (VST) counts was performed. This analysis highlights the presence of three samples with aberrant expression profiles (Fig. 3).    

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Create a DESeq2 object:
deseq_object  <- DESeq(dds_htseq,
                       test = "LRT",
                       reduced = ~caste + stage)  

## Transform gene-level counts using variance-stablising transformtion:
vsd <- vst(deseq_object,
             blind = FALSE)

## Save object:
saveRDS(object = vsd,
        file = "results/vsd_brain_treatment.rds")

## Calculate distance between samples based on transformed expression values:
sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$caste,
                                     vsd$treatment_group,
                                     vsd$stage,
                                     sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
```

```{r figure3, ref.label = "figure3", fig.width = 8, fig.height = 8, echo = FALSE, message=FALSE}
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

_**Fig. 3.** Heatmap displaying distance matrix of genes expressed within and across samples._   

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
#poisd <- PoissonDistance(t(counts(dds_htseq)))

#samplePoisDistMatrix <- as.matrix(poisd$dd)
#rownames(samplePoisDistMatrix) <- paste(dds_htseq$caste,
#                                        dds_htseq$treatment_group,
#                                        sep=" - " )
#colnames(samplePoisDistMatrix) <- NULL
```

```{r figure4, ref.label = "figure4", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
#pheatmap(samplePoisDistMatrix,
#         clustering_distance_rows = poisd$dd,
#         clustering_distance_cols = poisd$dd)
```

\newpage  

## Principal component analysis: Full dataset  

To examine further similarities in terms of expression profiles across samples, a principal component analysis using VST gene values was performed. The first principal component, which explains 55% of the variance in the dataset identifies the possibility of outliers with no clear difference explained by the first two components that separate caste. There was evidence that one of the treatments may be driving the pattern (Fig. 5). It was decided to rerun the PCA with the four samples separated by PC1 removed to provide more resolution on differences between samples    
```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Here set the two principal components that you want to compare:
## It can any combination between 1 and 22:
#first_pc <- 2
#second_pc <- 3
## Create function for plotting:
plotPCA.san <- function(object,
                        first_pc = first_pca,
                        second_pc = second_pca,
                        intgroup = "condition",
                        ntop = 1000,
                        returnData = FALSE) {
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
                                                     length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  #return(pca)
  percentVar <- pca$sdev ^ 2 / sum(pca$sdev ^ 2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1,
                 paste,
                 collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  ## Select the PCAs and percentVar that you like instead of 1 and 2
  d <- data.frame(PC1 = pca$x[, first_pc],
                  PC2 = pca$x[, second_pc],
                  group = group, 
                  intgroup.df,
                  name = colData(vsd)[,1])
  if (returnData) {
    attr(d, "percentVar") <- percentVar[first_pc:second_pc]
    return(d)
  }
    ggplot(data = d,
           aes_string(x = "PC1",
                      y = "PC2",
                      color = "group")) +
            geom_point(data = d,
                       aes(fill = group),
                       colour = "black",
                       pch = 21,
                       size = 3) +
            xlab(paste0("PC", first_pc, ": ",
                        round(percentVar[first_pc] * 100),
                        "% Variance")) +
            ylab(paste0("PC", second_pc, ": ",
                        round(percentVar[second_pc] * 100),
                        "% Variance")) +
            coord_fixed() +
            #geom_text_repel(size = 3) +
            theme_bw() +
            theme(axis.title = element_text(size = 12,
                                            face = "bold"),
                  axis.text = element_text(size = 10,
                                           face = "plain"),
                  legend.title = element_text(size = 12,
                                              face = "bold"),
                  legend.text = element_text(size = 12,
                                              face = "plain"),
                  legend.position = "top")
}
```

Scatterplots displaying combinations of the first four principal components (PCs) for:  
- Caste (Fig. 4a)  
- Treatmnent (Fig. 4b)    
- Stage (Fig. 4c)    

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
#group_plot + geom_text(aes(label=colnames(vsd)),vjust=2)
## Generate plot:
caste_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("caste"))
caste_plot_pc1_2 <- caste_plot_pc1_2 +
  scale_fill_discrete(name = "Caste")

caste_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("caste"))
caste_plot_pc1_3 <- caste_plot_pc1_3 +
  scale_fill_discrete(name = "Caste")

caste_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("caste"))
caste_plot_pc1_4 <- caste_plot_pc1_4 +
  scale_fill_discrete(name = "Caste")
```

```{r figure5, ref.label = "figure5", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
ggarrange(caste_plot_pc1_2,
          caste_plot_pc1_3,
          caste_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```

_**Fig. 4a.** Principal component analysis with first two principal components shown. Individuals are coloured by caste (queen = "red"; worker = "blue)._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plot:
group_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("treatment_group"))
group_plot_pc1_2 <- group_plot_pc1_2 +
  scale_fill_discrete(name = "Treatment")

group_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("treatment_group"))
group_plot_pc1_3 <- group_plot_pc1_3 +
  scale_fill_discrete(name = "Treatment")

group_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("treatment_group"))
group_plot_pc1_4 <- group_plot_pc1_4 +
  scale_fill_discrete(name = "Treatment")
```

```{r figure6, ref.label = "figure6", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
ggarrange(group_plot_pc1_2,
          group_plot_pc1_3,
          group_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```

_**Fig. 4b.** Principal component analysis with first two principal components shown. Individuals are coloured by treatment (control = "red"; inseminated group = "blue", dilutant group = "green", injection group = "purple")._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
stage_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("stage"))
stage_plot_pc1_2 <- stage_plot_pc1_2 +
  scale_fill_discrete(name = "Stage")

stage_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("stage"))
stage_plot_pc1_3 <- stage_plot_pc1_3 +
  scale_fill_discrete(name = "Stage")

stage_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("stage"))
stage_plot_pc1_4 <- stage_plot_pc1_4 +
  scale_fill_discrete(name = "Stage")
```

```{r figure7, ref.label = "figure7", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
ggarrange(stage_plot_pc1_2,
          stage_plot_pc1_3,
          stage_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```
_**Fig. 4c.** Principal component analysis with first two principal components shown. Individuals are coloured by treatment (two days = "red"; four days = "green"; eight days = "blue")_    

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Calculate principal components:
pca <- prcomp(t(assay(vsd)))

pca_df <- as.data.frame(pca$x)
## check the proportion of variance explained by each PC:
summary(pca)
```
\newpage  

## Principal component analysis: Reduced dataset  

Based on the cluster analyses using full samples (distance matrix Fig.3; PCA Fig. 4), there are certain samples (n = 3) that are masking more detailed understanding of substructure within the dataset based on expression profiles.  
Rerunning and plotting the output of the PCA with plots coloured by caste (Fig. 5a), treatment (Fig. 5b) and stage (Fig. 5C). PC1, which explained 30% of variance in the dataset largely separated individuals based on stage ovarian development with individuals collected at 8 days post-ovarian activation clearly separated from the others. PC2, which explained 17%, largely (but not completely) clustered queens away from workers. There was no clear pattern based on treatment.  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Identify outliers:
## Based on PCs 1 and 2, anything right of 10 on PC1 is likely an outlier:
samples_to_keep <- subset(x = pca_df,
                          PC1 < 50)
samples_to_keep <- row.names(samples_to_keep)

## Subset to only keep sampels of interest:
dds_htseq <- dds_htseq[, samples_to_keep]
  
## Recreate a DESeq2 object using these subsetted samples:
deseq_object  <- DESeq(dds_htseq,
                       test = "LRT",
                       reduced = ~caste + stage)  

## Transform gene-level counts using variance-stablising transformtion:
vsd <- vst(deseq_object,
             blind = FALSE)

## Generate plot:
caste_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("caste"))
caste_plot_pc1_2 <- caste_plot_pc1_2 +
  scale_fill_discrete(name = "Caste")

caste_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("caste"))
caste_plot_pc1_3 <- caste_plot_pc1_3 +
  scale_fill_discrete(name = "Caste")

caste_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("caste"))
caste_plot_pc1_4 <- caste_plot_pc1_4 +
  scale_fill_discrete(name = "Caste")
```

```{r figure8, ref.label = "figure8", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
ggarrange(caste_plot_pc1_2,
          caste_plot_pc1_3,
          caste_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```

_**Fig. 5a.** Principal component analysis with first two principal components shown. Individuals are coloured by caste (queen = "red"; worker = "blue)._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plot:
group_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("treatment_group"))
group_plot_pc1_2 <- group_plot_pc1_2 +
  scale_fill_discrete(name = "Treatment")

group_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("treatment_group"))
group_plot_pc1_3 <- group_plot_pc1_3 +
  scale_fill_discrete(name = "Treatment")

group_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("treatment_group"))
group_plot_pc1_4 <- group_plot_pc1_4 +
  scale_fill_discrete(name = "Treatment")
```

```{r figure9, ref.label = "figure9", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
ggarrange(group_plot_pc1_2,
          group_plot_pc1_3,
          group_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```

_**Fig. 5b.** Principal component analysis with first two principal components shown. Individuals are coloured by treatment (control = "red"; inseminated group = "blue", dilutant group = "green", injection = "purple")._  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
stage_plot_pc1_2 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 2,
            intgroup = c("stage"))
stage_plot_pc1_2 <- stage_plot_pc1_2 +
  scale_fill_discrete(name = "Stage")

stage_plot_pc1_3 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 3,
            intgroup = c("stage"))
stage_plot_pc1_3 <- stage_plot_pc1_3 +
  scale_fill_discrete(name = "Stage")

stage_plot_pc1_4 <- plotPCA.san(vsd,
                          first_pc = 1,
                          second_pc = 4,
            intgroup = c("stage"))
stage_plot_pc1_4 <- stage_plot_pc1_4 +
  scale_fill_discrete(name = "Stage")
```

```{r figure10, ref.label = "figure10", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
ggarrange(stage_plot_pc1_2,
          stage_plot_pc1_3,
          stage_plot_pc1_4,
          nrow = 1,
          ncol = 3,
          align = "hv",
          legend = "none")
```
_**Fig. 5c.** Principal component analysis with first two principal components shown. Individuals are coloured by treatment (two days = "red"; four days = "green"; eight days = "blue")_    

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Calculate principal components:
pca <- prcomp(t(assay(vsd)))

pca_df <- as.data.frame(pca$x)
## check the proportion of variance explained by each PC:
summary(pca)
```
\newpage  

## Differential expression analysis:  
As a preliminary examination of differences in gene expression due to treatment,
all samples, irrespective of caste and stage, were included in a differential expression 
analysis.

**Model 1:** To examine the effect of **treatment** on brain gene expression, we constructed a _full model_ consisting of:  
gene expression = ~treatment + caste + stage;   
and a _reduced model_ of:  
gene expression = ~caste + stage  

From this model, a total of **2554** genes were identified as significantly differentiated expressed (adjusted _p_ value < 0.05; Fig. 6) across treatments.  

Applying a more stringent threshold to examine specific differences between treatment groups for significant differential expression based on adjusted
_p_ value and a minimum of absolute log2 fold change of 1, our model identifies **115** (Fig. 7a)
significantly differentially expressed genes with the majority (n = **111**) having elevated 
expression in the brain of inseminated bees, which is significant deviation from equal outcome of expression profiles for differentially expressed genes (binomial test; *p* < 0.05).  

**Comparison of dilutant group and control bees**: Our model identifies **38** (Fig. 7b)
significantly differentially expressed genes with the majority (n = **33**) having elevated 
expression in the brain of bees inseminated only with dilutant.  

**Comparison of the "nothing" group and control bees**: Our model identifies **63** (Fig. 7c)
significantly differentially expressed genes with the majority (n = **40**) having elevated 
expression in the brain of bees treated only with the needle. A complication with this 
specific analysis is the low sample size of "nothing group", which consists of six individuals 
after the removal of outlier samples.  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Check names of comparisons:
resultsNames(deseq_object)

res <- results(deseq_object)
```

```{r figure11, ref.label = "figure11", fig.width = 8, fig.height = 5, echo = FALSE, message=FALSE}
## Plot distribution of adjusted p values:
hist(res$padj[res$baseMean > 1],
     breaks = 0:20/20,
     col = "grey50",
     border = "white",
     main = "Histogram of adjusted p-values for differentially expressed genes",
     xlab = "Adjusted p value")
```
_**Fig. 6** Histogram of adjiusted p values for differentially expressed genes between treatment groups._    
```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Extract results:
deseq_object_results <- results(deseq_object,
                                      name = "treatment_group_Insemination_group_vs_Control")

## Extract significant genes:
nrow(subset(x = deseq_object_results,
            padj < 0.05))

deseq_object_results$absL2FC <- abs(deseq_object_results$log2FoldChange)
deseq_object_results_sig     <- subset(x = deseq_object_results,
                                           padj < 0.05 &
                                           absL2FC > 1)
nrow(deseq_object_results_sig)

## Convert DESeq2 result object to a dataframe:
deseq_object_results <- as.data.frame(deseq_object_results)
deseq_object_results_sig <- as.data.frame(deseq_object_results_sig)

## Subset genes with evidence of significantly elevated expression in inseminated individuals:
deseq_object_insemination_sig_up     <- subset(x = deseq_object_results,
                                             padj < 0.05 &
                                            log2FoldChange >= 1)
nrow(deseq_object_insemination_sig_up)

## Subset genes with evidence of significantly reduced expression in inseminated individuals:
deseq_object_insemination_sig_down <- subset(x = deseq_object_results,
                                             padj < 0.05 &
                                             log2FoldChange <= -1)
nrow(deseq_object_insemination_sig_down)

## Plot volcano plot of:
## a) Log2FoldChange (gene expression) against:
## b) -log10 transformed p-value (significant threshold)
volcano_plot <- ggplot(deseq_object_results,
       aes(x = log2FoldChange,
           y = -log10(padj))) +
  xlim(-10, 10) +
  geom_point(alpha = 0.5) +
  geom_point(data = subset(deseq_object_insemination_sig_up,
                           log2FoldChange >= 1 &
                             padj < 0.05),
             aes(colour = "orange")) +
  geom_point(data = subset(deseq_object_insemination_sig_down,
                           log2FoldChange <= -1 &
                             padj < 0.05),
             aes(colour = "blue")) +
  geom_vline(xintercept = c(-1, 1),
             linetype = "dashed") +
  theme_bw() +
  theme(
    axis.title = element_text(face = "bold",
                              size = 12),
    legend.title = element_text(face = "bold",
                                size = 12),
    legend.text = element_text(face = "plain",
                                size = 12),
    legend.position = "top") +
  scale_colour_manual(values = c("blue",
                                 "orange"),
                      labels = c("Inseminated-reduced",
                                 "Inseminated-biased"),
                      name = "Status")
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
resultsNames(deseq_object)

## Extract results:
dilutant_object_results <- results(deseq_object,
                                      name = "treatment_group_Insemination_with_dilutent_vs_Control")

## Extract significant phenotype genes:
dilutant_object_results$absL2FC <- abs(dilutant_object_results$log2FoldChange)
dilutant_object_results_sig     <- subset(x = dilutant_object_results,
                                           padj < 0.05 &
                                           absL2FC > 1)
nrow(dilutant_object_results_sig)

## Convert DESeq2 result object to a dataframe:
dilutant_object_results <- as.data.frame(dilutant_object_results)
dilutant_object_results_sig <- as.data.frame(dilutant_object_results_sig)

## Subset genes with evidence of significantly elevated expression in inseminated individuals:
dilutant_object_insemination_sig_up     <- subset(x = dilutant_object_results,
                                             padj < 0.05 &
                                            log2FoldChange >= 1)
nrow(dilutant_object_insemination_sig_up)

## Subset genes with evidence of significantly reduced expression in inseminated individuals:
dilutant_object_insemination_sig_down <- subset(x = dilutant_object_results,
                                             padj < 0.05 &
                                             log2FoldChange <= -1)
nrow(dilutant_object_insemination_sig_down)

## Plot volcano plot of:
## a) Log2FoldChange (gene expression) against:
## b) -log10 transformed p-value (significant threshold)
dilutant_volcano_plot <- ggplot(dilutant_object_results,
       aes(x = log2FoldChange,
           y = -log10(padj))) +
  xlim(-10, 10) +
  geom_point(alpha = 0.5) +
  geom_point(data = subset(dilutant_object_insemination_sig_up,
                           log2FoldChange >= 1 &
                             padj < 0.05),
             aes(colour = "orange")) +
  geom_point(data = subset(dilutant_object_insemination_sig_down,
                           log2FoldChange <= -1 &
                             padj < 0.05),
             aes(colour = "blue")) +
  geom_vline(xintercept = c(-1, 1),
             linetype = "dashed") +
  theme_bw() +
  theme(
    axis.title = element_text(face = "bold",
                              size = 12),
    legend.title = element_text(face = "bold",
                                size = 12),
    legend.text = element_text(face = "plain",
                                size = 12),
    legend.position = "top") +
  scale_colour_manual(values = c("blue",
                                 "orange"),
                      labels = c("Dilutant-reduced",
                                 "Dilutant-biased"),
                      name = "Status")
```
```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
resultsNames(deseq_object)

## Extract results:
injection_object_results <- results(deseq_object,
                                      name = "treatment_group_Insemination_with_nothing_vs_Control")

## Extract significant phenotype genes:
injection_object_results$absL2FC <- abs(injection_object_results$log2FoldChange)
injection_object_results_sig     <- subset(x = injection_object_results,
                                           padj < 0.05 &
                                           absL2FC > 1)
nrow(injection_object_results_sig)

## Convert DESeq2 result object to a dataframe:
injection_object_results <- as.data.frame(injection_object_results)
injection_object_results_sig <- as.data.frame(injection_object_results_sig)

## Subset genes with evidence of significantly elevated expression in inseminated individuals:
injection_object_insemination_sig_up     <- subset(x = injection_object_results,
                                             padj < 0.05 &
                                            log2FoldChange >= 1)
nrow(injection_object_insemination_sig_up)

## Subset genes with evidence of significantly reduced expression in inseminated individuals:
injection_object_insemination_sig_down <- subset(x = injection_object_results,
                                             padj < 0.05 &
                                             log2FoldChange <= -1)
nrow(injection_object_insemination_sig_down)

## Plot volcano plot of:
## a) Log2FoldChange (gene expression) against:
## b) -log10 transformed p-value (significant threshold)
injection_volcano_plot <- ggplot(injection_object_results,
       aes(x = log2FoldChange,
           y = -log10(padj))) +
  xlim(-10, 10) +
  geom_point(alpha = 0.5) +
  geom_point(data = subset(injection_object_insemination_sig_up,
                           log2FoldChange >= 1 &
                             padj < 0.05),
             aes(colour = "orange")) +
  geom_point(data = subset(injection_object_insemination_sig_down,
                           log2FoldChange <= -1 &
                             padj < 0.05),
             aes(colour = "blue")) +
  geom_vline(xintercept = c(-1, 1),
             linetype = "dashed") +
  theme_bw() +
  theme(
    axis.title = element_text(face = "bold",
                              size = 12),
    legend.title = element_text(face = "bold",
                                size = 12),
    legend.text = element_text(face = "plain",
                                size = 12),
    legend.position = "top") +
  scale_colour_manual(values = c("blue",
                                 "orange"),
                      labels = c("injection-reduced",
                                 "injection-biased"),
                      name = "Status")
```

```{r figure12, ref.label = "figure12", fig.width = 8, fig.height = 10, echo = FALSE, message=FALSE, warning=FALSE}  
## Print to console:
ggarrange(volcano_plot,
          dilutant_volcano_plot,
          injection_volcano_plot,
          nrow = 3,
          ncol = 1,
          labels = c("A",
                     "B",
                     "C"))
```

_**Fig. 7.** Volcano plots displaying differences in gene expression across bumblebees differing in: a) insemination status; b) dilutant treatment; and c) injection treatment._   

```{r figure13, ref.label = "figure13", fig.width = 8, fig.height = 8, echo = FALSE, message=FALSE, warning=FALSE}  
## Examine also with a Euler plot:
## Compare overlap of differentially expressed genes:
venneu_treatment_comparison <- euler(combinations = list("Insemination up-DEGs" = row.names(deseq_object_insemination_sig_up),
                                                      "Insemination down-DEGs" = row.names(deseq_object_insemination_sig_down),
                                                      "Dilutant up-DEGs" = row.names(dilutant_object_insemination_sig_up),
                                                      "Dilutant down-DEGs" = row.names(dilutant_object_insemination_sig_down),
                                                      "Injection up-DEGs" = row.names(injection_object_insemination_sig_up),
                                                      "Injection down-DEGs" = row.names(injection_object_insemination_sig_down)),
                                     shape = "ellipse")

## Generate plot:
treatment_comparison_plot <- plot(venneu_treatment_comparison,
                               quantities = TRUE,
                               edges = TRUE,
                               labels = list(fontsize = 12))

## Print:
treatment_comparison_plot
         
## Save object:
saveRDS(object = deseq_object_results,
        file = "results/brain_deseq_object_results_treatment.rds")
```
_**Fig. 8.** Euler plot displaying differences and overlap in significantly differentially expressed genes for different treatment groups in comparison to control bees._   

## Interaction effect between caste and stage:  
**Model 2:** Given that we lack queen samples for the 'dilutant' and 'nothing' groups,
we cannot examine an interaction effect between caste and treatment unless we subset. We can however examine an interaction between caste and stage using the entire dataset:

Full model: gene expression = ~treatment + caste + stage + caste:stage 

We then construct a _reduced model_ consisting of:
gene expression = ~treatment + caste + stage  

Analysis identified **1668** genes as differentially expressed (adjusted _p_ value < 0.05) between stage depending on caste. This however could be a consequence of the increased number of worker samples, including more divergent groupings.  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Build DESeq2 dataset:
design(deseq_object) <- ~ caste +
                           stage +
                           treatment_group +
                           caste:stage

## Create a DESeq2 object:
deseq_object  <- DESeq(deseq_object,
                       test = "LRT",
                       reduced = ~caste +
                                  stage +
                                  treatment_group)

## Check names of comparisons:
resultsNames(deseq_object)
## Extract results:
deseq_object_results <- results(deseq_object,
                                      name = "casteWorker.stageStage_II")

## Extract significant phenotype genes:
deseq_object_results_sig     <- subset(x = deseq_object_results,
                                             padj < 0.05)
nrow(deseq_object_results_sig)

## Convert DESeq2 result object to a dataframe:
deseq_object_results <- as.data.frame(deseq_object_results)
deseq_object_results_sig <- as.data.frame(deseq_object_results_sig)
```
\newpage  
## Viusaliation of differentially expressed genes:
A related question - how similar are differences between control and inseminated workers and queens?

- First, we can visualise expression profiles for all individuals across samples using a heatmap for differentially expressed genes identified by Model 1 (Fig. 9).  
- Second, we subdivide the data into worker and queen datasets, run differential expression analysis and examine the number of shared differentially expressed genes, as well as their rank. Related to this, we can compare significantly enriched Gene Ontology terms.  

## Overlap analysis of differentially expressed genes:  
If we subset the dataset into workers and queens and examine differentially expressed genes with caste-restricted analyses, do we find the same differentially expressed genes between artifically inseminated and control bumblebees.  

Using subsetted worker samples for only 'control' and 'insemination treatment', **3266** genes were significantly differentially expressed. Applying a threshold of absolute log2 fold change of 1, we reduce our number of differentially expressed genes to **253** (Fig. 9A).  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## For this analysis, we will restrict to only 'control' and 'inseminated samples':
samples_info <- subset(x = samples_info,
                       treatment_group == "Control" | 
                         treatment_group == "Insemination_group")

## Create a function to run DESeq2:
run_caste_deseq = function(specific_caste){
             ## Subset caste samples:
             samples <- subset(x = samples_info,
                               caste == specific_caste)
             ## Create model containing stage and treatment group as explanatory variables.
             dds <- DESeqDataSetFromHTSeqCount(sampleTable = samples,
                                       directory = directory,
                                       design = ~stage +
                                                 treatment_group)
             ## Create a DESeq2 object:
             dds_object  <- DESeq(dds,
                                  test = "LRT",
                                  reduced = ~stage)
             ## Extract results:
             object_results <<- results(dds_object,
                                      name = "treatment_group_Insemination_group_vs_Control")
}

## Run differential expression analysis with workers:
run_caste_deseq(specific_caste = "Worker")
worker_results_df <- as.data.frame(object_results)
worker_results_df$absLog2FC <- abs(worker_results_df$log2FoldChange)
nrow(subset(x = worker_results_df,
            padj < 0.05))
nrow(subset(x = worker_results_df,
            padj < 0.05 &
            absLog2FC >= 1))

## Set a function for plotting a volcano plot:
plot_volcano <- function(input){
  ggplot(input,
       aes(x = log2FoldChange,
           y = -log10(padj))) +
  xlim(-10, 10) +
  geom_point(alpha = 0.5) +
  geom_point(data = subset(input,
                           padj < 0.05 &
                           log2FoldChange >= 1),
             aes(colour = "orange")) +
  geom_point(data = subset(input,
                           padj < 0.05 & 
                           log2FoldChange <= -1),
             aes(colour = "blue")) +
  geom_vline(xintercept = c(-1, 1),
             linetype = "dashed") +
  theme_bw() +
  theme(
    axis.title = element_text(face = "bold",
                              size = 12),
    legend.title = element_text(face = "bold",
                                size = 12),
    legend.text = element_text(face = "plain",
                                size = 12),
    legend.position = "top") +
  scale_colour_manual(values = c("blue",
                                 "orange"),
                      labels = c("Inseminated-reduced",
                                 "Inseminated-biased"),
                      name = "Status")
}

## Assign output to a unique object:
volcano_worker <- plot_volcano(input = worker_results_df)

## Save opbject:
saveRDS(object = worker_results_df,
        file = "results/brain_deseq_object_results_worker_treatment.rds")
```

Using all queen samples, **235** genes were differentially expressed between inseminated and control bees (Fig. 9B), which was reduced to **42** when applying a more stringent filter (Log2FC >= 1).  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Run differential expression analysis with queens:
run_caste_deseq(specific_caste = "Queen")
queen_results_df <- as.data.frame(object_results)
queen_results_df$absLog2FC <- abs(queen_results_df$log2FoldChange)
nrow(subset(x = queen_results_df,
            padj < 0.05))
nrow(subset(x = queen_results_df,
            padj < 0.05 &
            absLog2FC >= 1))

volcano_queen <- plot_volcano(input = queen_results_df)

saveRDS(object = queen_results_df,
        file = "results/brain_deseq_object_results_queen_treatment.rds")
```
```{r figure14, ref.label = "figure14", fig.width = 6, fig.height = 4, echo = FALSE, message=FALSE, warning=FALSE}
## Print to console:
ggarrange(volcano_worker,
          volcano_queen,
          nrow = 2,
          ncol = 1,
          align = "hv",
          labels = c("A",
                     "B"),
          legend = "none")
```

_**Fig. 9.** Volcano plots displaying differences in gene expression of brain sampled from (A) workers and (B) queens differing in insemination status. Dashed vertical lines highlight log2FC of 1 and -1, respectively. Blue dots indicate genes with significantly reduced expression (adjusted p value < 0.05 & log2FC <= -1) in brains from inseminated individuals compared to control. Orange dots indicate genes with significantly elevated expression (adjusted p value < 0.05 & log2FC >= 1) in brains from inseminated individuals compared to control._     

\newpage

## Euler plot:  
To visualise similarities in terms of differential expression across bumblebee castes in response to insemination, we used a euler plot. First, in response to insemination, workers and queens share **23** differentially expressed genes, which consists of approximately 10% and 52% of worker and queen differentially expressed genes, respectively (Fig. 10).  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
worker_object_results_sig <- subset(worker_results_df,
                                    padj < 0.05 &
                                      absLog2FC >= 1)
queen_object_results_sig <- subset(queen_results_df,
                                    padj < 0.05 &
                                      absLog2FC >= 1)

## Compare overlap of differentially expressed genes:
venneu_tissue_comparison <- euler(combinations = list("Worker DEGs" = row.names(worker_object_results_sig),
                                                      "Queen DEGs" = row.names(queen_object_results_sig)))

## Generate plot:
tissue_comparison_plot <- plot(venneu_tissue_comparison,
                               quantities = TRUE,
                               edges = TRUE,
                               labels = list(fontsize = 12))
```

```{r figure15, ref.label = "figure15", fig.width = 4, fig.height = 2, fig.align = 'center', echo = FALSE, message=FALSE, warning=FALSE}
tissue_comparison_plot
```

_**Fig. 10.** Euler plot of differentially expressed genes between inseminated and control individuals for each caste._

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Check rank:
worker_object_results_ordered <- worker_results_df[order(worker_results_df$pvalue), ]
worker_object_results_ordered$rank <- 1:ncol(worker_object_results_ordered)
queen_object_results_ordered <- queen_results_df[order(queen_results_df$pvalue), ]
queen_object_results_ordered$rank <- 1:ncol(queen_object_results_ordered)

## Create a dataframe matched between the worker and queen dataframe:
worker_object_results_reordered <- worker_object_results_ordered[match(row.names(queen_object_results_ordered),
                                    row.names(worker_object_results_ordered)), ]

head(worker_object_results_reordered)
head(queen_object_results_ordered)

tmp <- as.data.frame(cbind(row.names(worker_object_results_reordered),
                           worker_object_results_reordered$log2FoldChange,
                           worker_object_results_reordered$padj,
                           queen_object_results_ordered$log2FoldChange,
                           queen_object_results_ordered$padj))

tmp$worker_rank <- worker_object_results_reordered$rank
tmp$queen_rank <- row.names(tmp)

tmp$V2 <- as.numeric(as.character(unlist(tmp$V2)))
tmp$V3 <- as.numeric(as.character(unlist(tmp$V3)))
tmp$V4 <- as.numeric(as.character(unlist(tmp$V4)))
tmp$V5 <- as.numeric(as.character(unlist(tmp$V5)))
```

To further examine similarities in terms of gene expression in response to insemination,
we ranked genes based on log2 fold change between inseminated and control bees for 
each caste and performed a Pearson's coefficient correlation based on log2 fold change 
identifying a weak (but significant) positive correlation ( _R_ = 0.12, *p* < 0.05; Fig. 11).

```{r figure16, ref.label = "figure16", fig.width = 6, fig.height = 4, fig.align = 'center', echo = FALSE, message=FALSE, warning=FALSE}
## Generate a scatterplot:
ggplot(data = tmp,
       aes(x = V2,
           y = V4)) +
  geom_smooth(method = "lm") +
  xlab(label = "Worker: insemination vs control - log2FC") +
  ylab(label = "Queen: insemination vs control - log2FC") +
  geom_point(alpha = 0.5) +
  geom_point(data = subset(x = tmp,
                           V3 <= 0.05 &
                             abs(V2) >= 1),
                           colour = "red",
             alpha = 0.5) +
  geom_point(data = subset(x = tmp,
                           V5 <= 0.05 &
                           abs(V2) >= 1),
                           colour = "yellow",
             alpha = 0.5) +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 12)) +
  stat_cor(method = "pearson")
```

_**Fig. 11.** Scatterplot of correlation between log2 fold change in gene expression for inseminated and control for both bumblebee workers and queens. Each individual point represents a gene with coloured points representing genes with statistically significant differences in gene expression between treatment groups. Different colours represent different castes (red = worker, yellow = queen, orange = overlap)._  

## Gene Ontology enrichment analysis:
As a complementary approach to examine overlap in terms of higher processes related
to differential gene expression, rank-based Gene Ontology enrichment analyses were performed
for both worker and queen datasets, respectively. For each dataset, log2 fold change for
inseminated bees versus control was used as a ranking measure and converted into a range
from 1 to 0 to allow for implementation of a KS test in topGO, which analysed each GO term
subcategory (biological process (BP); molecular function (MF); and celluar component (CC))
in turn. Correction for multiple testing was performed using the Benjamini-Hochberg method. 
For both worker and queen datasets, significantly enriched GO terms (n = 390 and 141, respectively)  
were examined for overlap using a Euler plot identifying 35 enriched GO terms shared amongst
both the worker and queen significantly differentially expressed genes providing further insight
into similar changes in both castes in terms of gene expression in response to insemination.  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Read in enriched go terms:
go_terms_worker_bp <- read.table(file = "data/output_20_brain_ks_log2FC_worker_classic/BP_sig_new.tmp",
                              header = TRUE)
go_terms_worker_bp$GO.ID <- as.character(unlist(go_terms_worker_bp$GO.ID))
nrow(go_terms_worker_bp)
go_terms_queen_bp <- read.table(file = "data/output_20_brain_ks_log2FC_queen_classic/BP_sig_new.tmp",
                              header = TRUE)
nrow(go_terms_queen_bp)
go_terms_queen_bp$GO.ID <- as.character(unlist(go_terms_queen_bp$GO.ID))
go_terms_worker_cc <- read.table(file = "data/output_20_brain_ks_log2FC_worker_classic/CC_sig_new.tmp",
                              header = TRUE)
go_terms_worker_cc$GO.ID <- as.character(unlist(go_terms_worker_cc$GO.ID))
go_terms_queen_cc <- read.table(file = "data/output_20_brain_ks_log2FC_queen_classic/CC_sig_new.tmp",
                              header = TRUE)
go_terms_queen_cc$GO.ID <- as.character(unlist(go_terms_queen_cc$GO.ID))
go_terms_worker_mf <- read.table(file = "data/output_20_brain_ks_log2FC_worker_classic/MF_sig_new.tmp",
                              header = TRUE)
go_terms_worker_mf$GO.ID <- as.character(unlist(go_terms_worker_mf$GO.ID))
go_terms_queen_mf <- read.table(file = "data/output_20_brain_ks_log2FC_queen_classic/MF_sig_new.tmp",
                              header = TRUE)
go_terms_queen_mf$GO.ID <- as.character(unlist(go_terms_queen_mf$GO.ID))
## Extract and compare significantly enriched GO terms:
go_term_comparison <- euler(combinations = list("Worker GO terms" = c(subset(x = go_terms_worker_bp,
                                                                            classic_ks_adjusted < 0.05)$GO.ID,
                                                                      subset(x = go_terms_worker_cc,
                                                                             classic_ks_adjusted < 0.05)$GO.ID,
                                                                      subset(x = go_terms_worker_mf,
                                                                             classic_ks_adjusted < 0.05)$GO.ID),
                                                      "Queen GO terms" = c(subset(x = go_terms_queen_bp,
                                                                                  classic_ks_adjusted < 0.05)$GO.ID,
                                                                          subset(x = go_terms_queen_cc,
                                                                                 classic_ks_adjusted < 0.05)$GO.ID,
                                                                           subset(x = go_terms_queen_mf,
                                                                                  classic_ks_adjusted < 0.05)$GO.ID)))
## Generate plot:
go_term_comparison_plot <- plot(go_term_comparison,
                               quantities = TRUE,
                               edges = TRUE,
                               labels = list(fontsize = 12))
```

```{r figure17, ref.label = "figure17", fig.width = 4, fig.height = 2, fig.align = 'center',, echo = FALSE, message=FALSE, warning=FALSE}
go_term_comparison_plot
```

_**Fig. 12.** Euler plot of enriched GO terms between inseminated and control individuals for each caste._  

\newpage

## Weighted gene co-expression network analysis:

Weighted gene co-expression network analysis was performed using the R package, WCGNA. 
The WGCNA identified 15 modules, of which one module ("darkmagenta") was significantly positively correlated with treatment (Fig. 13). However, the module was also positively correlated with caste and stage suggesting that other factors may explain gene expression profiles better.  

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Load saved R objects from the WGCNA analysis:
load(file = "data/brain-03-input_for_labelled_heatmap.RData")
#load(file = "data/spermatheca-04-input_for_membership_scatterplot.RData")
```

```{r figure18, ref.label = "figure18", fig.width = 6, fig.height = 6, fig.align = 'center',, echo = FALSE, message=FALSE, warning=FALSE}
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = exp_moduleTraitCor,
             xLabels = names(samples_information),
             yLabels = names(exp_MEs),
             ySymbols = names(exp_MEs),
             colorLabels = FALSE,
             colors = blueWhiteRed(50),
             textMatrix = exp_textMatrix,
             setStdMargins = FALSE,
             cex.text = 0.5,
             zlim = c(-1,1),
             main = paste("Module-category relationships"))
```

_**Fig. 13.** Heatmap displaying correlative relationships between WGCNA modules and categories (caste, stage, and treatment) of interest._ 

\newpage
**Table 1:** Sample information

```{r, echo = FALSE, message = FALSE}
row.names(samples_info) <- gsub(pattern = ".ReadsPerGene.out.tab",
                                      replacement = "",
                                      row.names(samples_info))
knitr::kable(samples_info[, 3:6], "pipe")
```
