---
author: "Joe Colgan"
title: "**Transcriptomic analysis of control and inseminated spermatheca of _Bombus terrestris_ workers and queens**"
output:
  pdf_document: default
  html_document: default
fig_width: 4
fig_height: 4
fontsize: 20pt
---

## Introduction
The purpose of the present analysis is to examine similarities and differences in terms of gene expression associated with the spermatheca of _Bombus terrestris_ workers and queens. For the present study, control (unmated) and artificially inseminated bumblebees were collected and spermatheca removed at three different time points (two, four and eight days post-treatment).

1. Load libraries:

```{r, message = FALSE}
#install.packages("BiocManager") 
#BiocManager::install("WGCNA")
#install.packages("igraph")
library(WGCNA)
library(DESeq2)
library(igraph)
library(ggcorrplot)
```

2. Load input data:

The expression data is contained in the file LiverFemale3600.csv that comes with this tutorial. After starting an R session, we load the requisite packages and the data, after appropriately setting the working directory:

```{r, message = FALSE}
# The following setting is important, do not omit.
options(stringsAsFactors = FALSE)

## Load experimental data:
data <- readRDS(file = "input/vsd_spermatheca_treatment.rds")
data_counts <- assay(data)
```

3. In addition to expression data, the data files contain extra information about the surveyed probes we do not need. One can inspect larger data frames such as femData by invoking R data editor via fix(femData). The expression data set contains 135 samples. Note that each row corresponds to a gene and column to a sample or auxiliary information. We now remove the auxiliary data and transpose the expression data for further analysis.

```{r, message=FALSE}
## We need a transposed dataframe consisting of gene per column and sample per row:
transposed_data <- t(data_counts)
```

4. Checking data for excessive missing values and identification of outlier microarray samples

We first check for genes and samples with too many missing values:

```{r, message = FALSE}
## Run for experimental dataset:
transposed_gsg <- goodSamplesGenes(transposed_data, verbose = 3)
transposed_gsg$allOK
```

If the last statement returns TRUE, all genes have passed the cuts. If not, we remove the offending genes and samples from the data:

```{r, message = FALSE}
## Run for experimental:
if (!transposed_gsg$allOK)
{
  # Optionally, print the gene and sample names that were removed:
  if (sum(!transposed_gsg$goodGenes)>0)
     printFlush(paste("Removing genes:", paste(names(transposed_data)[!transposed_gsg$goodGenes], collapse = ", ")));
  if (sum(!transposed_gsg$goodSamples)>0)
     printFlush(paste("Removing samples:", paste(rownames(transposed_data)[!transposed_gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  transposed_data = transposed_data[transposed_gsg$goodSamples, transposed_gsg$goodGenes]
}
```

5. Next we cluster the samples (in contrast to clustering genes that will come later) to see if there are any obvious outliers.

```{r, message = FALSE}
## Run for experimental samples:
exp_sampleTree = hclust(dist(transposed_data), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)
#pdf(file = "Plots/sampleClustering.pdf", width = 12, height = 9);
par(cex = 0.6);
par(mar = c(0,4,2,0))
plot(exp_sampleTree, main = "Sample clustering to detect outliers",
     sub = "",
     xlab = "",
     cex.lab = 1.5,
     cex.axis = 1.5,
     cex.main = 2)
```

It appears there is one outlier (sample F2_221, see Fig. 1). One can remove it by hand, or use an automatic approach. Choose a height cut that will remove the offending sample, say 15 (the red line in the plot), and use a branch cut at that height.

```{r, message = FALSE}
## For the experimental dataset, there is an issue with sample 37:
# Plot a line to show the cut
#abline(h = 15, col = "red");
# Determine cluster under the line
clust = cutreeStatic(exp_sampleTree,
                     cutHeight = 150,
                     minSize = 10)
table(clust)
# clust 1 contains the samples we want to keep.
keepSamples = (clust==1)
transposed_data_filt = transposed_data[keepSamples, ]
nGenes = ncol(transposed_data_filt)
nSamples = nrow(transposed_data_filt)
```

The variable datExpr now contains the expression data ready for network analysis.  

6 Loading trait data
We now read in the trait data and match the samples for which they were measured to the expression samples.

```{r, message = FALSE}
## Read in sample information:
samples_information <- read.table(file = "data/sample_info_for_wgcna.txt",
                                  header = FALSE)
## Remove tissue:
samples_information$V3 <- NULL
row.names(samples_information) <- samples_information$V1
samples_information$V1 <- NULL
## Update column names:
colnames(samples_information) <- c("caste",
                                   "age",
                                   "treatment")
## Remove outlier:
samples_information <- subset(samples_information,
                              row.names(samples_information) %in% row.names(transposed_data_filt))
```

We now have the expression data in the variable datExpr, and the corresponding clinical traits in the variable datTraits. Before we continue with network construction and module detection, we visualize how the clinical traits relate to the sample dendrogram.  

```{r, message = FALSE}
## Run for experimental dataset:
# Re-cluster samples
exp_sampleTree2 = hclust(dist(transposed_data_filt),
                         method = "average")
# Convert traits to a color representation: white means low, red means high, grey means missing entry
exp_traitColors = numbers2colors(samples_information, signed = FALSE);

exp_sampleTree2$labels <- gsub(pattern = ".ReadsPerGene.out.tab",
                               replacement = "",
                               exp_sampleTree2$labels)

# Plot the sample dendrogram and the colors underneath:
pdf(file = "results/sample_dendrogram_trait_heatmap.pdf",
    height = 6,
    width = 8)
plotDendroAndColors(exp_sampleTree2,
                    exp_traitColors,
                  groupLabels = gsub(pattern = ".ReadsPerGene.out.tab",
                  replacement = "",
                  names(samples_information)),
                  main = "Sample dendrogram and trait heatmap")
dev.off()

saveRDS(object = exp_sampleTree2, file = "results/exp_sampleTree2_spermatheca.rds")
saveRDS(object = exp_traitColors, file = "results/exp_traitColors_spermatheca.rds")
```

In the plot, shown in Fig. 2, white means a low value, red a high value, and grey a missing entry. The last step is to save the relevant expression and trait data for use in the next steps of the tutorial. 

7 Step-by-step network construction and module detection
7.1 Choosing the soft-thresholding power: analysis of network topology
Constructing a weighted gene network entails the choice of the soft thresholding power β to which co-expression similarity is raised to calculate adjacency [1]. The authors of [1] have proposed to choose the soft thresholding power based on the criterion of approximate scale-free topology. We refer the reader to that work for more details; here we illustrate the use of the function pickSoftThreshold that performs the analysis of network topology and aids the user in choosing a proper soft-thresholding power. The user chooses a set of candidate powers (the function provides suitable default values), and the function returns a set of network indices that should be inspected, for example as follows:  

```{r, message = FALSE}
## For the experimental set:
# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12,
                        to = 20,
                        by = 2))
# Call the network topology analysis function
exp_sft = pickSoftThreshold(transposed_data_filt,
                            powerVector = powers,
                            verbose = 5)
# Plot the results:
sizeGrWindow(9, 5)
par(mfrow = c(1,2));
cex1 = 0.9;
# Scale-free topology fit index as a function of the soft-thresholding power
plot(exp_sft$fitIndices[,1],
     -sign(exp_sft$fitIndices[,3])*exp_sft$fitIndices[,2],
     xlab = "Soft Threshold (power)",
     ylab = "Scale Free Topology Model Fit,signed R^2",
     type = "n",
     main = paste("Scale independence"));
text(exp_sft$fitIndices[,1],
     -sign(exp_sft$fitIndices[,3])*exp_sft$fitIndices[,2],
     labels = powers,
     cex = cex1,
     col = "red");
# this line corresponds to using an R^2 cut-off of h
abline(h = 0.90,
       col = "red")
# Mean connectivity as a function of the soft-thresholding power
plot(exp_sft$fitIndices[,1],
     exp_sft$fitIndices[,5],
     xlab = "Soft Threshold (power)",
     ylab = "Mean Connectivity",
     type = "n",
     main = paste("Mean connectivity"))
text(exp_sft$fitIndices[,1],
     exp_sft$fitIndices[,5],
     labels = powers,
     cex = cex1,
     col = "red")
```

7.2 Co-expression similarity and adjacency
We now calculate the adjacencies, using the soft thresholding power 6:

```{r, message = FALSE}
## For experimental dataset:
softPower = 6;
exp_adjacency = adjacency(transposed_data_filt,
                          power = softPower);
```

7.3 Topological Overlap Matrix (TOM)
To minimize effects of noise and spurious associations, we transform the adjacency into Topological Overlap Matrix, and calculate the corresponding dissimilarity:

```{r, message = FALSE}
## Experimental dataset:
# Turn adjacency into topological overlap
exp_TOM = TOMsimilarity(exp_adjacency);
diss_exp_TOM = 1-exp_TOM
```

7.4 Clustering using TOM
We now use hierarchical clustering to produce a hierarchical clustering tree (dendrogram) of genes. Note that we use the function hclust that provides a much faster hierarchical clustering routine than the standard hclust function.  

```{r, message = FALSE}
## Experimental data:
exp_geneTree = hclust(as.dist(diss_exp_TOM),
                      method = "average");
# Plot the resulting clustering tree (dendrogram)
sizeGrWindow(12,9)
plot(exp_geneTree,
     xlab = "",
     sub = "",
     main = "Gene clustering on TOM-based dissimilarity",
     labels = FALSE,
     hang = 0.04);
```

The clustering dendrogram plotted by the last command is shown in Figure 2. In the clustering tree (dendrogram), each leaf, that is a short vertical line, corresponds to a gene. Branches of the dendrogram group together densely interconnected, highly co-expressed genes. Module identification amounts to the identification of individual branches (”cutting the branches off the dendrogram”). There are several methods for branch cutting; our standard method is the Dynamic Tree Cut from the package dynamicTreeCut. The next snippet of code illustrates its use.  

```{r, message = FALSE}
## Experimental set:
# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 30;
# Module identification using dynamic tree cut:
exp_dynamicMods = cutreeDynamic(dendro = exp_geneTree,
                                distM = diss_exp_TOM,
              deepSplit = 2,
              pamRespectsDendro = FALSE,
              minClusterSize = minModuleSize);
table(exp_dynamicMods)
```

The function returned 22 modules labeled 1–22 largest to smallest. Label 0 is reserved for unassigned genes. The above command lists the sizes of the modules. We now plot the module assignment under the gene dendrogram:

```{r, message=FALSE}
## Experimental data:
# Convert numeric lables into colors
exp_dynamicColors = labels2colors(exp_dynamicMods)
table(exp_dynamicColors)
# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)
plotDendroAndColors(exp_geneTree,
                    exp_dynamicColors,
                    "Dynamic Tree Cut",
                    dendroLabels = FALSE,
                    hang = 0.03,
                    addGuide = TRUE,
                    guideHang = 0.05,
                    main = "Gene dendrogram and module colors")
```

7.5 Merging of modules whose expression profiles are very similar
The Dynamic Tree Cut may identify modules whose expression profiles are very similar. It may be prudent to merge such modules since their genes are highly co-expressed. To quantify co-expression similarity of entire modules, we calculate their eigengenes and cluster them on their correlation  

```{r, message = FALSE}
## Experimental dataset:
# Calculate eigengenes
exp_MEList = moduleEigengenes(transposed_data_filt,
                              colors = exp_dynamicColors)
exp_MEs = exp_MEList$eigengenes
# Calculate dissimilarity of module eigengenes
exp_MEDiss = 1-cor(exp_MEs);
# Cluster module eigengenes
exp_METree = hclust(as.dist(exp_MEDiss),
                    method = "average");
# Plot the result
sizeGrWindow(7, 6)
plot(exp_METree,
     main = "Clustering of module eigengenes",
     xlab = "",
     sub = "")
```

We choose a height cut of 0.25, corresponding to correlation of 0.75, to merge (see Fig. 4):

```{r, message = FALSE}
## for the experimental dataset:
exp_MEDissThres = 0.25
# Plot the cut line into the dendrogram
# abline(h=exp_MEDissThres, col = "red")
# Call an automatic merging function
exp_merge = mergeCloseModules(transposed_data_filt,
                              exp_dynamicColors,
                              cutHeight = exp_MEDissThres,
                              verbose = 3)
# The merged module colors
exp_mergedColors = exp_merge$colors;
# Eigengenes of the new merged modules:
exp_mergedMEs = exp_merge$newMEs;
```

To see what the merging did to our module colors, we plot the gene dendrogram again, with the original and merged module colors underneath (Figure 5).

```{r, message = FALSE}
sizeGrWindow(12, 9)
#pdf(file = "Plots/geneDendro-3.pdf", wi = 9, he = 6)
plotDendroAndColors(exp_geneTree, cbind(exp_dynamicColors,
                                        exp_mergedColors),
                  c("Dynamic Tree Cut", "Merged dynamic"),
                  dendroLabels = FALSE,
                  hang = 0.03,
                  addGuide = TRUE,
                  guideHang = 0.05)
```

In the subsequent analysis, we will use the merged module colors in mergedColors. We save the relevant variables for use in subsequent parts of the tutorial:

```{r, message = FALSE}
## Experimental dataset:
# Rename to moduleColors
exp_moduleColors = exp_mergedColors
# Construct numerical labels corresponding to the colors
exp_colorOrder = c("grey", standardColors(50));
exp_moduleLabels = match(exp_moduleColors,
                         exp_colorOrder)-1;
exp_MEs = exp_mergedMEs;
dir.create(path = "results")
# Save module colors and labels for use in subsequent parts
save(exp_MEs,
     exp_moduleLabels,
     exp_moduleColors,
     exp_geneTree,
     file = "results/spermatheca-02-networkConstruction-stepByStep.RData")
```

8. Relating modules to external traits
8.1 Quantifying module–trait associations
In this analysis we would like to identify modules that are significantly associated with the measured  traits. Since we already have a summary profile (eigengene) for each module, we simply correlate eigengenes with external traits and look for the most significant associations:

```{r, message = FALSE}
# Define numbers of genes and samples
exp_nGenes = ncol(transposed_data_filt);
exp_nSamples = nrow(transposed_data_filt);
# Recalculate MEs with color labels
exp_MEs0 = moduleEigengenes(transposed_data_filt,
                            exp_moduleColors)$eigengenes
exp_MEs = orderMEs(exp_MEs0)
exp_moduleTraitCor = cor(exp_MEs,
                         samples_information,
                         use = "p");
exp_moduleTraitPvalue = corPvalueStudent(exp_moduleTraitCor,
                                         exp_nSamples);
```

Since we have a moderately large number of modules and traits, a suitable graphical representation will help in reading the table. We color code each association by the correlation value:  

```{r, message = FALSE}
## Experimental dataset:
sizeGrWindow(10,6)
# Will display correlations and their p-values
exp_textMatrix = paste(signif(exp_moduleTraitCor, 2), "\n(",
                        signif(exp_moduleTraitPvalue, 1), ")", sep = "");
dim(exp_textMatrix) = dim(exp_moduleTraitCor)
# Display the correlation values within a heatmap plot:
pdf(file = "results/module-trait_relationship_heatmap.pdf",
    height = 6,
    width = 8)
par(mar = c(6, 8.5, 3, 3));
labeledHeatmap(Matrix = exp_moduleTraitCor,
             xLabels = names(samples_information),
             yLabels = names(exp_MEs),
             ySymbols = names(exp_MEs),
             colorLabels = FALSE,
             colors = blueWhiteRed(50),
             textMatrix = exp_textMatrix,
             setStdMargins = FALSE,
             cex.text = 0.5,
             zlim = c(-1,1),
             main = paste("Module-trait relationships"))
dev.off()

# Save module colors and labels for use in subsequent parts
save(exp_MEs,
     exp_moduleTraitCor,
     samples_information,
     exp_textMatrix,
     file = "results/spermatheca-03-input_for_labelled_heatmap.RData")
```

The resulting color-coded table is shown in Fig. 1.
The analysis identifies the several significant module–trait associations. We will concentrate on weight as the trait of interest.  

8.2 Gene relationship to trait and important modules: Gene Significance and Module Membership
We quantify associations of individual genes with our trait of interest (treatment) by defining Gene Significance GS as (the absolute value of) the correlation between the gene and the trait. For each module, we also define a quantitative measure of module membership MM as the correlation of the module eigengene and the gene expression profile. This allows us to quantify the similarity of all genes on the array to every module.

```{r, message = FALSE}
# Define variable treatment containing the treatment column of samples information
treatment = as.data.frame(samples_information$treatment);
names(treatment) = "treatment"
# names (colors) of the modules
exp_modNames = substring(names(exp_MEs), 3)
exp_geneModuleMembership = as.data.frame(cor(transposed_data_filt,
                                         exp_MEs,
                                         use = "p"));
exp_MMPvalue = as.data.frame(corPvalueStudent(as.matrix(exp_geneModuleMembership),
                                          exp_nSamples));

names(exp_geneModuleMembership) = paste("MM",
                                        exp_modNames,
                                        sep="");
names(exp_MMPvalue) = paste("p.MM",
                            exp_modNames,
                            sep = "");
exp_geneTraitSignificance = as.data.frame(cor(transposed_data_filt,
                                          treatment,
                                          use = "p"));
exp_GSPvalue = as.data.frame(corPvalueStudent(as.matrix(exp_geneTraitSignificance),
                                          exp_nSamples));
names(exp_geneTraitSignificance) = paste("GS.",
                                         names(treatment),
                                         sep = "");
names(exp_GSPvalue) = paste("p.GS.",
                            names(treatment),
                            sep = "")
```

8.3 Intramodular analysis: identifying genes with high GS and MM
Using the GS and MM measures, we can identify genes that have a high significance for treatment as well as high module membership in interesting modules. As an example, we look at the brown module that has the highest association with weight. We plot a scatterplot of Gene Significance vs. Module Membership in the brown module:  

```{r, message = FALSE}
exp_module = "yellow"
exp_column = match(exp_module,
                   exp_modNames);
exp_moduleGenes = exp_moduleColors == exp_module;
sizeGrWindow(7, 7);
par(mfrow = c(1,1));
verboseScatterplot(abs(exp_geneModuleMembership[exp_moduleGenes, exp_column]),
                   abs(exp_geneTraitSignificance[exp_moduleGenes, 1]),
                 xlab = paste("Module Membership in", exp_module, "module"),
                 ylab = "Gene significance for treatment",
                 main = paste("Module membership vs. gene significance\n"),
                 cex.main = 1.2,
                 cex.lab = 1.2,
                 cex.axis = 1.2,
                 col = exp_module)

# Save module colors and labels for use in subsequent parts
save(exp_geneModuleMembership,
     exp_moduleGenes,
     exp_module,
     exp_column,
     exp_geneTraitSignificance,
     file = "results/spermatheca-04-input_for_membership_scatterplot_yellow.RData")

## Generate a combined plot:
pdf(file = "results/module-trait_relationship_heatmap.pdf",
    height = 6,
    width = 12)
par(mar = c(6, 8.5, 3, 3));
labeledHeatmap(Matrix = exp_moduleTraitCor[1:5,],
             xLabels = names(samples_information),
             yLabels = names(exp_MEs)[1:5],
             ySymbols = names(exp_MEs)[1:5],
             colorLabels = FALSE,
             colors = blueWhiteRed(50),
             textMatrix = exp_textMatrix[1:5,],
             setStdMargins = FALSE,
             cex.text = 1,
             zlim = c(-1,1),
             main = paste(""))
dev.off()

verboseScatterplot(abs(exp_geneModuleMembership[exp_moduleGenes, exp_column]),
                   abs(exp_geneTraitSignificance[exp_moduleGenes, 1]),
                 xlab = paste("Module Membership in", exp_module, "module"),
                 ylab = "Gene significance for treatment",
                 main = paste("Module membership vs. gene significance\n"),
                 cex.main = 1.2,
                 cex.lab = 1.2,
                 cex.axis = 1.2,
                 col = exp_module)
```


3.d Summary output of network analysis results
We have found modules with high association with our trait of interest, and have identified their central players by the Module Membership measure. We now merge this statistical information with gene annotation and write out a file that summarizes the most important results and can be inspected in standard spreadsheet software such as MS Excel or Open Office Calc. Our expression data are only annotated by probe ID names: the command 
will return all probe IDs included in the analysis. Similarly, the second command below 
will return probe IDs belonging to the brown module. To facilitate interpretation of the results, we use a probe annotation file provided by the manufacturer of the expression arrays to connect probe IDs to gene names and universally recognized identification numbers (Entrez codes).

````{r, message = FALSE}
transposed_data_filt <- as.data.frame(transposed_data_filt)
names(transposed_data_filt)
red_module_genes <- names(transposed_data_filt)[exp_moduleColors=="red"]
length(red_module_genes)
write(x = red_module_genes,
      file = "results/red_module_genes.txt")

yellow_module_genes <- names(transposed_data_filt)[exp_moduleColors=="yellow"]
length(yellow_module_genes)
write(x = yellow_module_genes,
      file = "results/yellow_module_genes.txt")

pink_module_genes <- names(transposed_data_filt)[exp_moduleColors=="pink"]
length(pink_module_genes)
write(x = pink_module_genes,
      file = "results/pink_module_genes.txt")

turquoise_module_genes <- names(transposed_data_filt)[exp_moduleColors=="turquoise"]
length(turquoise_module_genes)
write(x = turquoise_module_genes,
      file = "results/turquoise_module_genes.txt")

grey_module_genes <- names(transposed_data_filt)[exp_moduleColors=="grey"]
length(grey_module_genes)
write(x = grey_module_genes,
      file = "results/grey_module_genes.txt")

black_module_genes <- names(transposed_data_filt)[exp_moduleColors=="black"]
length(black_module_genes)
write(x = black_module_genes,
      file = "results/black_module_genes.txt")
```
Read back in enriched Gene Ontology terms for:
- Genes belonging to the red module
- Genes belonging to the yellow module

```{r, message = FALSE}
## Read in Gene Ontology terms enriched for resident genes:
go_term_dir <- "../2022_gene_ontology_analysis/run_reps/results/bterrestris_dmel"
## Load ALH output:
red_bp_terms <- read.table(file = paste(go_term_dir,
                                        "/output_20_red_module_fisher_weight01",
                                        "/BP_sig.tmp",
                                              sep = ""),
                           header = TRUE)
red_bp_terms$log10 <- -log10(red_bp_terms$weight_fisher)
red_bp_terms$category <- "Red module genes"

## Subset by significance:
yellow_bp_terms <- read.table(file = paste(go_term_dir,
                                          "/output_20_yellow_module_fisher_weight01",
                                          "/BP_sig.tmp",
                                              sep = ""),
                           header = TRUE)
yellow_bp_terms$log10 <- -log10(yellow_bp_terms$weight_fisher)
yellow_bp_terms$category <- "Yellow module genes"

## Generate plot:
combined_df <- as.data.frame(rbind(red_bp_terms,
                                   yellow_bp_terms))
combined_df_unique <- subset(combined_df,
                             !duplicated(combined_df$GO.ID))

## Reannotate:
combined_df_unique$Term <- gsub(pattern = "_",
                                replacement = " ",
                                combined_df_unique$Term)

## Create a column containing amend terms:
combined_df_unique$amended_terms <- paste(combined_df_unique$Term,
                                          " ",
                                          "(",
                                          combined_df_unique$Annotated,
                                          ")",
                                          sep = "")

## Relevel category factors:
combined_df_unique$category <- factor(combined_df_unique$category,
                                      levels = c("Red module genes",
                                                 "Yellow module genes"))

## Generate plot:
plot <- ggbarplot(combined_df_unique,
                  x = "amended_terms",
                  y = "log10",
                  position = position_dodge(0.1),
                  fill = "category",
                  color = NULL,
                  palette = "jco",
                  sort.val = "desc",
                  sort.by.groups = TRUE,
                  ylab = "-log10(p)",
                  xlab = "Enriched Gene Ontology term",
                  legend.title = "Gene categories",
                  x.text.angle = 90,
                  lab.col = "black",
                  lab.size = 4,
                  lab.vjust = 0.5,
                  lab.hjust = 1,
                  legend = "top",
                  rotate = FALSE,
                  ggtheme = theme_minimal())

## Make font bigger and bold:
plot <- plot +
  scale_fill_manual(values = c("red",
                               "gold")) +
                scale_y_continuous(expand = c(0, 0)) +
                theme(axis.text = element_text(size = 10),
                      axis.title.x = element_text(size = 12,
                                                  face = "bold"),
                      axis.title.y = element_text(size = 12,
                                                  face = "bold"),
                      axis.text.y = element_text(size = 10,
                                                 face = "bold"),
                      axis.text.x = element_text(size = 10),
                      legend.position = "top",
                      legend.title = element_text(size = 10,
                                                  face = "bold")) +
                expand_limits(y = 10) +
                geom_hline(yintercept = 1.301,
                           linetype = "dashed",
                           colour = "black") +
                geom_hline(yintercept = -1.301,
                           linetype = "dashed",
                           colour = "black")

## Generate a correlation plot using ggcorrplot:
correlation_plot <- ggcorrplot(exp_moduleTraitCor,
           lab = TRUE) +
  theme(axis.text = element_text(size = 10,
                                  face = "bold"),
        axis.text.x = element_text(size = 10,
                                   colour = c("turquoise",
                                              "pink",
                                              "black",
                                              "red",
                                              "gold",
                                              "darkgrey"),
                                  face = "bold"),
        legend.title = element_text(size = 10,
                                  face = "bold"))

## Generate a combined plot:
ggarrange(correlation_plot,
          plot,
          ncol = 1,
          nrow = 2,
          labels = c("a",
                     "b"),
          heights = c(1,1.5))

## Save image:
ggsave(file = "results/combined_cor_go_term_plot.pdf",
       height = 12,
       width = 8,
       dpi = 600)
```

5 Visualization of networks within R 5.a Visualizing the gene network
One way to visualize a weighted network is to plot its heatmap, Fig. 1. Each row and column of the heatmap correspond to a single gene. The heatmap can depict adjacencies or topological overlaps, with light colors denoting low adjacency (overlap) and darker colors higher adjacency (overlap). In addition, the gene dendrograms and module colors are plotted along the top and left side of the heatmap. The package provides a convenient function to create such network plots; Fig. 1 was created using the following code. This code can be executed only if the network was calculated using a single-block approach (that is, using the 1-step automatic or the step-by-step tutorials). If the networks were calculated using the block-wise approach, the user will need to modify this code to perform the visualization in each block separately. The modification is simple and we leave it as an exercise for the interested reader.

```{r, message = FALSE}
# Calculate topological overlap anew: this could be done more efficiently by saving the TOM
# calculated during module detection, but let us do it again here.
dissTOM = 1-TOMsimilarityFromExpr(transposed_data_filt, power = 6);
# Transform dissTOM with a power to make moderately strong connections more visible in the heatmap
plotTOM = dissTOM^7;
# Set diagonal to NA for a nicer plot
diag(plotTOM) = NA;
# Call the plot function
sizeGrWindow(9,9)
TOMplot(plotTOM,
        exp_geneTree,
        exp_moduleColors,
        main = "Network heatmap plot, all genes")
```

Note that the generating the heatmap plot may take a substantial amount of time. It is possible to restrict the number of genes to speed up the plotting; however, the gene dendrogram of a subset of genes will often look different from the gene dendrogram of all genes. In the following example we restrict the number of plotted genes to 400:

```{r, message = FALSE}
nSelect = 400
# For reproducibility, we set the random seed
set.seed(10);
select = sample(nGenes, size = nSelect);
selectTOM = dissTOM[select, select];
# There’s no simple way of restricting a clustering tree to a subset of genes, so we must re-cluster.
selectTree = hclust(as.dist(selectTOM),
                    method = "average")
selectColors = exp_moduleColors[select];
# Open a graphical window
sizeGrWindow(9,9)
# Taking the dissimilarity to a power, say 10, makes the plot more informative by effectively changing
# the color palette; setting the diagonal to NA also improves the clarity of the plot
plotDiss = selectTOM^7;
diag(plotDiss) = NA;
TOMplot(plotDiss, selectTree, selectColors, main = "Network heatmap plot, selected genes")
```

5.b Visualizing the network of eigengenes
It is often interesting to study the relationships among the found modules. One can use the eigengenes as represen- tative profiles and quantify module similarity by eigengene correlation. The package contains a convenient function plotEigengeneNetworks that generates a summary plot of the eigengene network. It is usually informative to add a clinical trait (or multiple traits) to the eigengenes to see how the traits fit into the eigengene network:  
```{r, message = FALSE}
# Recalculate module eigengenes
MEs = moduleEigengenes(transposed_data_filt,
                       exp_moduleColors)$eigengenes
# Isolate weight from the clinical traits
treatment = as.data.frame(samples_information$treatment);
names(treatment) = "treatment"
# Add the weight to existing module eigengenes
MET = orderMEs(cbind(MEs, treatment))
# Plot the relationships among the eigengenes and the trait
sizeGrWindow(5,7.5);
par(cex = 0.9)
plotEigengeneNetworks(MET, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2), cex.lab = 0.8, xLabelsAngle
= 90)
```

The function produces a dendrogram of the eigengenes and trait(s), and a heatmap of their relationships. To split the dendrogram and heatmap plots, we can use the following code

```{r, message = FALSE}
# Plot the dendrogram
sizeGrWindow(6,6);
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene dendrogram", marDendro = c(0,4,2,0),
                    plotHeatmaps = FALSE)
# Plot the heatmap matrix (note: this plot will overwrite the dendrogram plot)
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
                    plotDendrograms = FALSE, xLabelsAngle = 90)
```

Fig. 2 shows the output of the above code. The eigengene dendrogram and heatmap identify groups of correlated eigengenes termed meta-modules. For example, the dendrogram indicates that red, brown and bluw modules are highly related; their mutual correlations are stronger than their correlations with weight. On the other hand, the salmon module, which is also significantly correlated with weight, is not part of the same meta-module as the red,
   
brown and blue modules, at least if meta-modules are defined as tight custers of modules (for example, modules with a correlation of eigengenes of at least 0.5).

```{r, message = FALSE}
# Recalculate topological overlap if needed
TOM = TOMsimilarityFromExpr(transposed_data_filt,
                            power = 6);
# Read in the annotation file
#annot = read.csv(file = "GeneAnnotation.csv");
# Select modules
modules = c("yellow", "red");
# Select module probes
probes = names(transposed_data_filt)
inModule = is.finite(match(exp_moduleColors, modules));
modProbes = probes[inModule];
#modGenes = annot$gene_symbol[match(modProbes, annot$substanceBXH)];
# Select the corresponding Topological Overlap
modTOM = TOM[inModule, inModule];

dimnames(modTOM) = list(modProbes, modProbes)
# Export the network into edge and node list files Cytoscape can read
cyt = exportNetworkToCytoscape(modTOM,
  edgeFile = paste("CytoscapeInput-edges-", paste(modules, collapse="-"), ".txt", sep=""),
  nodeFile = paste("CytoscapeInput-nodes-", paste(modules, collapse="-"), ".txt", sep=""),
  weighted = TRUE,
  threshold = 0.02,
  nodeNames = modProbes,
  altNodeNames = c(red_module_genes, yellow_module_genes),
  nodeAttr = exp_moduleColors[inModule]);
```

Igraph analysis:

```{r, message = FALSE}
## Subset "yellow genes"
## Let us transpose back:
data_filt <- t(transposed_data_filt)

## Subset genes of interest:
data_filt_yellow <- subset(x = data_filt,
                           row.names(data_filt) %in% yellow_module_genes)

## Remove 'unmapped' and 'unambiguous' counts:
data_filt_yellow <- data_filt_yellow[-c(1,2), ]

# Create a graph adjacency based on correlation distances between genes in  pairwise fashion.
g <- graph.adjacency(
  as.matrix(as.dist(cor(t(data_filt_yellow), method="pearson"))),
  mode="undirected",
  weighted=TRUE,
  diag=FALSE
)

# Simplfy the adjacency object
g <- simplify(g, remove.multiple=TRUE, remove.loops=TRUE)

# Colour negative correlation edges as blue
E(g)[which(E(g)$weight<0)]$color <- "darkblue"

# Colour positive correlation edges as red
E(g)[which(E(g)$weight>0)]$color <- "darkred"

# Convert edge weights to absolute values
E(g)$weight <- abs(E(g)$weight)

# Remove edges below absolute Pearson correlation 0.8
g <- delete_edges(g, E(g)[which(E(g)$weight<0.8)])

# Remove any vertices remaining that have no edges
g <- delete.vertices(g, degree(g)==0)

# Assign names to the graph vertices (optional)
V(g)$name <- V(g)$name

# Change shape of graph vertices
V(g)$shape <- "sphere"

# Change colour of graph vertices
V(g)$color <- "skyblue"

# Change colour of vertex frames
V(g)$vertex.frame.color <- "white"

# Scale the size of the vertices to be proportional to the level of expression of each gene represented by each vertex
# Multiply scaled vales by a factor of 10
scale01 <- function(x){(x-min(x))/(max(x)-min(x))}
vSizes <- (scale01(apply(data_filt_yellow, 1, mean)) + 1.0) * 10

# Amplify or decrease the width of the edges
edgeweights <- E(g)$weight * 2.0

# Convert the graph adjacency object into a minimum spanning tree based on Prim's algorithm
mst <- mst(g, algorithm="prim")

# Plot the tree object
plot(
  mst,
  layout=layout.fruchterman.reingold,
  edge.curved=TRUE,
  vertex.size=vSizes,
  vertex.label.dist=-0.5,
  vertex.label.color="black",
  asp=FALSE,
  vertex.label.cex=0.6,
  edge.width=edgeweights,
  edge.arrow.mode=0,
  main="Genes of the yellow module")

## Choose top hub for each module:
chooseTopHubInEachModule(transposed_data_filt,
                         colorh = exp_moduleColors)
```
