---
author: "Joe Colgan"
title: "**Gene Ontology term enrichment analysis of control and inseminated spermatheca of _Bombus terrestris_ workers and queens**"
output:
  pdf_document: default
  html_document: default
fig_width: 4
fig_height: 4
fontsize: 20pt
---

## Introduction
The purpose of the present analysis is to examine similarities and differences in terms of gene expression associated with the spermatheca of _Bombus terrestris_ workers and queens. For the present study, control (unmated) and artificially inseminated bumblebees were collected and spermatheca removed at three different time points (two, four and eight days post-treatment).

This script is for gene ontology (GO) enrichment analysis of differentiallly expressed genes to explore the biological processes and molecular functions associated with DE genes, using the R package 'TopGO' (https://www.bioconductor.org/packages/3.7/bioc/vignettes/topGO/inst/doc/topGO.pdf).
Twp input files are required for the running of the test:  
1) A genelist file:  
The genelist file is a tab-delimited file containing two columns:  
Column 1: Locus (contains gene or transcript name of interest).  
Column 2: Rank value of interest (e.g. p-values or log fold changes).  
Rank values were generated using the R package, DESeq2, and implemented using code in ```gene_level_3_treatments.Rmd```.  

2) A GO database file:  
The GO database file is a tab-delimited file containing two columns:  
Column 1: Locus (contains gene or transcript name of interest).  
Column 2: Comma separated GO terms (e.g. GO:0000001, GO:0000002, etc.).  

This script will prepare data for GO analysis and create a 'TopGO object' from which enrichment tests can be performed to explore GO terms significantly enriched within the dataset. 
This script outputs a results table of significantly enriched GO terms.

*It should be run once per treatment, and will produce 3 output files each time it is run*

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("topGO")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        avebiocLite(pkgs = lib)
        library(lib, character.only = TRUE)
    }
}
```

## Step One: Input files, define output and objects for running topGO: 
Load in genelist and database files.  

```{r}
## Step One:
## Define input:
input <- "data/output_go_term_database.tmp"

## GO annotations
gene_to_go_mapping_file <- input

## file of significant genes (2 column file: i.e. gene id and pvalue) file:
deg_ranked_file <- readRDS(file = "input/fat_body_deseq_object_results_worker_treatment.rds")
deg_ranked_file <- as.data.frame(cbind(row.names(deg_ranked_file),
                                       deg_ranked_file$log2FoldChange))

## Define node size:
node_size <- 50
## Database:
database <- "bterrestris_dmel"
## Output directory:
output_directory <- paste("results/",
                          database,
                          "/output_",
                          node_size,
                          "_fat_body_ks_log2FC_worker_classic",
                          sep = "")
if (file.exists(output_directory)) {
  stop("The output directory:", output_directory, ", already exists",
       "Let's avoid overwriting")
} else {
  dir.create(output_directory,
             recursive = TRUE)
}
```

Read in input file:

```{r, message = FALSE}
## Read in information
colnames(deg_ranked_file) <- c("locus",
                               "log2FC")
genes_ranked <- deg_ranked_file[complete.cases(deg_ranked_file), ]
genes_ranked$log2FC <- as.numeric(as.character(unlist(genes_ranked$log2FC)))

## Sort by adjusted p value:
genes_ranked_sorted <- genes_ranked[order(-genes_ranked$log2FC), ]
range01 <- function(x){(x-min(x))/(max(x)-min(x))}
genes_ranked_sorted$range <- rev(range01(genes_ranked_sorted$log2FC))
## Remove na values:
## NA values are present for certain genes without expression, such as tRNAs:
genes_ranked_sorted_filtered <- subset(x      = genes_ranked_sorted,
                                   subset = !is.na(range))
## Read in GO annotations:
gene_to_go_mapping <- readMappings(file = gene_to_go_mapping_file)

## Convert into topgo's genelist format:
topgo_genelist        <- genes_ranked_sorted_filtered$range
names(topgo_genelist) <- genes_ranked_sorted_filtered$locus
# Define a cut-off for running fisher's exact test:
cutoff_for_top_fivepercent <- quantile(x = topgo_genelist,
                                       probs = 0.05)
```

## Steps Two and Three: Create topGO Object & run tests for GO term enrichment

We create a topGO object for each GO term

We perform two statistical tests:
1. A ks test using the topGO 'weight01' algorithm
2. A Fisher's exact test using the topGO 'weight01' algoritm

We combine the output of each test. 
We filter out enriched terms.

We do this for each of the three GO categories (ie. Biological process, Molcular Function, Cellular Component):

```{r}
for (go_category in c("BP", "MF", "CC")) {
  # STEP TWO
  ## Build the GOdata object in topGO
  my_go_data <- new("topGOdata",
                    description = paste("GOtest", go_category, sep = "_"),
                    ontology    = go_category,
                    geneSel     = function(x) {
                                   # fails to run without this
                                     return(x <= cutoff_for_top_fivepercent)
                                   },
                    allGenes    = topgo_genelist,
                    gene2GO     = gene_to_go_mapping,
                    annot       = annFUN.gene2GO,
                    nodeSize    = node_size)
  # STEP THREE
  ## Calculate ks test using 'weight01' algorithm:
  result_classic_ks <- runTest(object    = my_go_data,
                              algorithm = "classic",
                              statistic = "ks")
  ## Combine results from statistical tests:
  result_classic_output <- GenTable(object        = my_go_data,
                                   classic_ks     = result_classic_ks,
                                   #weight_fisher = result_weight_fisher,
                                   orderBy       = "classic_ks",
                                   topNodes      = length(score(result_classic_ks)))
  ## Correct ks test for multiple testing:
  result_classic_output$classic_ks <- as.numeric(na.omit(result_classic_output$classic_ks))
  result_classic_output$classic_ks_adjusted <- p.adjust(p = result_classic_output$classic_ks,
                                                      method = c("BH"))
  result_classic_output_sig <- subset(result_classic_output,
                                     Significant > Expected &
                                       Expected > 1 &
                                      classic_ks < 0.05)
  
  ## Write to output:
  write.table(x         = result_classic_output_sig,
              file      = file.path(output_directory,
                                    paste(go_category,
                                          "sig.tsv",
                                          sep = "_")),
              row.names = FALSE,
              sep       = "\t",
              quote = FALSE)
}
```

Run lintr:

```{r, message = FALSE}
## Check style:
lintr::lint(file = "go_enrichment_analysis.Rmd")
```
